{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "**Table of contents**<a id='toc0_'></a>    \n",
    "- 1. [IMPORTS](#toc1_)    \n",
    "- 2. [DEFINE CONSTANT VARIABLES](#toc2_)    \n",
    "- 3. [FX: `build_TE_df_from_gff3`](#toc3_)    \n",
    "- 4. [IMPORT CHROMOSOME FILES AND GENERATE CHROMOSOME DATAFRAMES](#toc4_)    \n",
    "  - 4.1. [FXS: Get chromosome information](#toc4_1_)    \n",
    "    - 4.1.1. [`get_chr_file`](#toc4_1_1_)    \n",
    "    - 4.1.2. [`fix_chr_names`](#toc4_1_2_)    \n",
    "    - 4.1.3. [`get_relative_positions`](#toc4_1_3_)    \n",
    "    - 4.1.4. [`make_chr_df_from_fasta`](#toc4_1_4_)    \n",
    "    - 4.1.5. [Define chromosome lengths dataframes](#toc4_1_5_)    \n",
    "  - 4.2. [IMPORT TE FILES AND CREATE TE DATAFRAMES](#toc4_2_)    \n",
    "- 5. [SIZE FILTERING](#toc5_)    \n",
    "  - 5.1. [FX: `create_small_range_dataframe`](#toc5_1_)    \n",
    "  - 5.2. [Generate small range dataframes](#toc5_2_)    \n",
    "  - 5.3. [FX: `make_gff3`](#toc5_3_)    \n",
    "  - 5.4. [STATS: percent of genome that is TE](#toc5_4_)    \n",
    "- 6. [GRAPHS](#toc6_)    \n",
    "  - 6.1. [graphing variables](#toc6_1_)    \n",
    "  - 6.2. [FX: `one_x_axis`](#toc6_2_)    \n",
    "  - 6.3. [ALL TE HISTOGRAMS](#toc6_3_)    \n",
    "    - 6.3.1. [FX: `make_ultimate_hist_onegen` (DNA and retro for N2, CB, VC2010 or KimCB)](#toc6_3_1_)    \n",
    "    - 6.3.2. [FX: `make_ultimate_hist_allgen` (all genomes for DNA or retro)](#toc6_3_2_)    \n",
    "  - 6.4. [SPECIFIC TE FAMILY HISTOGRAMS](#toc6_4_)    \n",
    "    - 6.4.1. [FX: `make_spec_family_ult_hist`](#toc6_4_1_)    \n",
    "    - 6.4.2. [FX: `make_spec_family_ult_hist_allgenomes`](#toc6_4_2_)    \n",
    "    - 6.4.3. [FX: `track_down_hist_peaks`](#toc6_4_3_)    \n",
    "    - 6.4.4. [FX: `make_every_fam_grid`](#toc6_4_4_)    \n",
    "  - 6.5. [TE COUNTS BAR GRAPH](#toc6_5_)    \n",
    "    - 6.5.1. [FX: `make_te_counts_perfam`](#toc6_5_1_)    \n",
    "    - 6.5.2. [FX: `make_class_counts_allgenomes`](#toc6_5_2_)    \n",
    "  - 6.6. [TE SIZES GRAPHS](#toc6_6_)    \n",
    "    - 6.6.1. [boxplot of all sizes on one plot](#toc6_6_1_)    \n",
    "    - 6.6.2. [FX: `broken_axis_box` ](#toc6_6_2_)    \n",
    "    - 6.6.3. [FX: `make_boxplot`](#toc6_6_3_)    \n",
    "    - 6.6.4. [FX: `make_split_violin`](#toc6_6_4_)    \n",
    "    - 6.6.5. [FX: `make_box_allgen`](#toc6_6_5_)    \n",
    "\n",
    "<!-- vscode-jupyter-toc-config\n",
    "\tnumbering=true\n",
    "\tanchor=true\n",
    "\tflat=false\n",
    "\tminLevel=1\n",
    "\tmaxLevel=6\n",
    "\t/vscode-jupyter-toc-config -->\n",
    "<!-- THIS CELL WILL BE REPLACED ON TOC UPDATE. DO NOT WRITE YOUR TEXT IN THIS CELL -->"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# 1. <a id='toc1_'></a>[IMPORTS](#toc0_)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 80,
   "id": "a22cb473",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2023-01-13T23:59:45.173327Z",
     "start_time": "2023-01-13T23:59:40.901708Z"
    }
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/var/folders/60/m3wm021x08vbp7kk3s0t94c40000gn/T/ipykernel_61597/2728266925.py:1: DeprecationWarning: Importing display from IPython.core.display is deprecated since IPython 7.14, please import from IPython display\n",
      "  from IPython.core.display import display, HTML\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<style>.container { width:100% !important; }</style>"
      ],
      "text/plain": [
       "<IPython.core.display.HTML object>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "from IPython.core.display import display, HTML\n",
    "display(HTML(\"<style>.container { width:100% !important; }</style>\"))\n",
    "from matplotlib import cm\n",
    "import seaborn as sns\n",
    "sns.set(style=\"white\")\n",
    "import pandas as pd\n",
    "from matplotlib.colors import ListedColormap, LinearSegmentedColormap\n",
    "from matplotlib.patches import Patch\n",
    "import csv\n",
    "import io\n",
    "from Bio import SeqIO\n",
    "import pickle\n",
    "from IPython.display import display, Markdown\n",
    "import matplotlib.pyplot as plt\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# 2. <a id='toc2_'></a>[DEFINE CONSTANT VARIABLES](#toc0_)\n",
    "- includes variables for colors, order lists, chromosome name lists etc that are used throughout the notebook"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "5be04c11",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2023-01-13T23:59:45.199914Z",
     "start_time": "2023-01-13T23:59:45.190369Z"
    }
   },
   "outputs": [],
   "source": [
    "# file path base\n",
    "filebase = '/Users/coraalbers/Documents/TEUlt/'\n",
    "\n",
    "# retro\n",
    "retro_cmap = ListedColormap(['#8B4F98','#3969AC','#F2B701','#E73F74','#80BA5A'])\n",
    "retro_color_dict = {'Copia':'#3969AC', 'ERV':'#80BA5A', 'Gypsy':'#8B4F98', 'LINE':'#F2B701', 'SINE':'#E73F74'}\n",
    "retro_color_list = ['#8B4F98', '#3969AC', '#F2B701', '#E73F74', '#80BA5A']\n",
    "\n",
    "\n",
    "# dna\n",
    "colorblind_list = sns.color_palette(palette = 'colorblind', as_cmap = True)\n",
    "dna_cmap = ListedColormap(colorblind_list)\n",
    "\n",
    "dna_color_dict = {'CMC':'#027e73', 'hAT':'de8f05', 'Helitron':'#fbafe4', 'MITE':'#ca9161', 'Novosib':'#949494', 'Sola':'#cc78bc', 'Tc1-Mariner':'#d55e00', 'Zator':'#0173b2'}\n",
    "dna_color_list = ['#0173b2', '#de8f05', '#027e73', '#d55e00', '#cc78bc', '#949494', '#fbafe4', '#ca9161']\n",
    "\n",
    "# orders\n",
    "dna_order = ['Zator', 'hAT', 'CMC', 'Tc1/Mariner', 'Sola', 'Novosib', 'Helitron', 'MITEs']\n",
    "retro_order = ['Gypsy', 'Copia', 'LINE', 'SINE', 'ERV']\n",
    "\n",
    "# chrom lists\n",
    "N2chromosomes = ['N2_chrI', 'N2_chrII', 'N2_chrIII', 'N2_chrIV', 'N2_chrV', 'N2_chrX']\n",
    "CBchromosomes = ['CB_chrI', 'CB_chrII', 'CB_chrIII', 'CB_chrIV', 'CB_chrV', 'CB_chrX']\n",
    "kimCBchromosomes = ['kCB_chrI', 'kCB_chrII', 'kCB_chrIII', 'kCB_chrIV', 'kCB_chrV', 'kCB_chrX']\n",
    "kimCBchromosomes_num = ['CP038187.1','CP038188.1','CP038189.1','CP038190.1','CP038191.1','CP038192.1' ]\n",
    "vc2010chromosomes = ['vc_chrI', 'vc_chrII', 'vc_chrIII', 'vc_chrIV', 'vc_chrV', 'vc_chrX']\n",
    "vc2010chromosomes_num = ['UNSB01000001.1','UNSB01000002.1','UNSB01000003.1','UNSB01000004.1','UNSB01000005.1',]\n",
    "chromosomes = ['Chr I', 'Chr II', 'Chr III', 'Chr IV', 'Chr V', 'X Chr']\n",
    "chromosomes_1 = ['I_1', 'II_1', 'III_1', 'IV', 'V_1', 'X_1']\n",
    "chromosomes_roman = ['I', 'II', 'III', 'IV', 'V', 'X']\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# 3. <a id='toc3_'></a>[FX: `build_TE_df_from_gff3`](#toc0_)\n",
    "- import gff3, calculate TE sizes and add in seq records, create dataframe, sort dataframe first by chromosome name and then start position\n",
    "- make sure to check the names of your chromosomes in your gff3 file match the names in the `gff_chrom_id` list in the function so that they are replaced with the TEUlt names correctly"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "id": "ca9af1fe",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2023-01-13T23:59:45.231326Z",
     "start_time": "2023-01-13T23:59:45.219111Z"
    }
   },
   "outputs": [],
   "source": [
    "def build_TE_df_from_gff3(gff3_file: str):\n",
    "    \n",
    "    gff_chrom_id = ['seq1', 'seq2', 'seq3', 'seq4', 'seq5', 'seq6']\n",
    "    chromosomes = ['Chr I', 'Chr II', 'Chr III', 'Chr IV', 'Chr V', 'X Chr']\n",
    "#     chromosomes = ['chrI', 'chrII', 'chrIII', 'chrIV', 'chrV', 'chrX']\n",
    "    \n",
    "    gff_list = []\n",
    "    \n",
    "    with open(gff3_file, newline = '') as lines:                #open gff file, take as tab delimited, create huge list that is the gff file\n",
    "#         \n",
    "#         line_reader = csv.reader(x.replace('\\0', '') for x in lines)\n",
    "        line_reader = csv.reader(lines, delimiter='\\t')\n",
    "        for line in line_reader:\n",
    "#             print(line)\n",
    "            gff_list.append(line)\n",
    "\n",
    "    gff_data = []\n",
    "    \n",
    "    for i in range(len(gff_list)):\n",
    "        \n",
    "#         print(i)\n",
    "        temp = [ gff_list[i][0], int(gff_list[i][3]), int(gff_list[i][4]), gff_list[i][5], gff_list[i][6] ] #chromsome, start, end, score, strand...\n",
    "        \n",
    "        size = int(gff_list[i][4]) - int(gff_list[i][3]) ##get TE size (stop_pos - start_pos) \n",
    "        temp.append(size)\n",
    "        \n",
    "        class_start = gff_list[i][8].find('(') +1\n",
    "        class_end = gff_list[i][8].find(')')\n",
    "        \n",
    "        class_info = gff_list[i][8][class_start:class_end].split(',') #find transposon class info/descriptions \n",
    "        \n",
    "        if len(class_info) == 3: ##if not a helitron...\n",
    "            \n",
    "            family = class_info[0]\n",
    "            sub_class = class_info[1]\n",
    "            main_class = class_info[2]\n",
    "\n",
    "\n",
    "            temp.append(family)\n",
    "            temp.append(sub_class)\n",
    "            temp.append(main_class) ##add these three class/ID fields to the temp line\n",
    "            \n",
    "        elif len(class_info) == 2: #if a helitron...\n",
    "            \n",
    "            family = class_info[0]\n",
    "            sub_class = '...'\n",
    "            main_class = class_info[1]\n",
    "\n",
    "\n",
    "            temp.append(family)\n",
    "            temp.append(sub_class)\n",
    "            temp.append(main_class) ##add these three class/ID fields to the temp line\n",
    "            \n",
    "        \n",
    "        gff_data.append(temp) ##append each \"line\" to data list of lists\n",
    "      \n",
    "    gff_df = pd.DataFrame(gff_data, columns=[\"Chromosome\", \"Start\", \"Stop\", \"Score:Probability\", \"Strand\", 'Size', 'Family', 'Subclass', 'Class']) #turn list of lists into dataframe\n",
    "    \n",
    "    for i in range(6):\n",
    "        \n",
    "        gff_df['Chromosome'] = gff_df['Chromosome'].replace(gff_chrom_id[i], chromosomes[i]) ##replace TransposonUltimate chromosome names with something better\n",
    "        \n",
    "    gff_df_sorted = gff_df.sort_values(by=['Chromosome', 'Start'], axis=0, ascending=True, ignore_index=True) ##sort by chromosome name first, then starting position within each chromosome, reindex rows...\n",
    "    \n",
    "    return gff_df_sorted, gff_list"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# 4. <a id='toc4_'></a>[IMPORT CHROMOSOME FILES AND GENERATE CHROMOSOME DATAFRAMES](#toc0_)\n",
    "## 4.1. <a id='toc4_1_'></a>[FXS: Get chromosome information](#toc0_)\n",
    "### 4.1.1. <a id='toc4_1_1_'></a>[`get_chr_file`](#toc0_)\n",
    "- import chr file and make df\n",
    "- this function is specific to the chr_lengths.txt file included, and could also be easily altered to accept more or less genomes within the chr_lengths.txt file (ex. 4 genomes instead of 2)\n",
    "### 4.1.2. <a id='toc4_1_2_'></a>[`fix_chr_names`](#toc0_)\n",
    "- changes chromosome names to the names used by TEUlt\n",
    "- change the `old_chr_names` list within the function to the names of your chromosomes\n",
    "\n",
    "### 4.1.3. <a id='toc4_1_3_'></a>[`get_relative_positions`](#toc0_)\n",
    "- calculates the relative start positions of transposons as a fraction of the length of the genome using the chromosome lengths from chr_df\n",
    "- allows for comparison of position of TEs between chromosomes with different lengths\n",
    "\n",
    "### 4.1.4. <a id='toc4_1_4_'></a>[`make_chr_df_from_fasta`](#toc0_)\n",
    "- makes the needed chromosome dataframe from a genome's fasta file (separates out chromosome sequences and gets the chromosome, start and stop positions)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "id": "2fa19366",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2023-01-13T23:59:45.259356Z",
     "start_time": "2023-01-13T23:59:45.252239Z"
    }
   },
   "outputs": [],
   "source": [
    "def get_chr_file(chr_lengths_file: str):\n",
    "    chr_lines_list = []\n",
    "    with open(chr_lengths_file, newline = '') as chr_lines:\n",
    "        chr_line_reader = csv.reader(chr_lines, delimiter='\\t')\n",
    "        for line in chr_line_reader:\n",
    "            chr_lines_list.append(line)\n",
    "            \n",
    "    N2_chr_list = []\n",
    "    CB_chr_list = []\n",
    "    for i in range(len(chr_lines_list)):\n",
    "        \n",
    "        if chr_lines_list[i][0] == 'Chr':\n",
    "            continue\n",
    "            \n",
    "        elif chr_lines_list[i][3] == 'Libuda N2':\n",
    "            chr_lines_list[i][2] = int(chr_lines_list[i][2])\n",
    "            N2_chr_list.append(chr_lines_list[i])\n",
    "            \n",
    "        elif chr_lines_list[i][3] == 'Libuda CB':\n",
    "            chr_lines_list[i][2] = int(chr_lines_list[i][2])\n",
    "            CB_chr_list.append(chr_lines_list[i])\n",
    "            \n",
    "    N2_chr_df = pd.DataFrame(N2_chr_list, columns = ['chr', 'start', 'end', 'strain'])\n",
    "    \n",
    "    CB_chr_df = pd.DataFrame(CB_chr_list, columns = ['chr', 'start', 'end', 'strain'])\n",
    "            \n",
    "    return N2_chr_df, CB_chr_df\n",
    "\n",
    "def fix_chr_names(dataframe):\n",
    "    \n",
    "    old_chr_names = ['I', 'II', 'III', 'IV', 'V', 'X']\n",
    "\n",
    "    for index, row in dataframe.iterrows():\n",
    "        if dataframe.at[index, 'chr'] == old_chr_names[0]:\n",
    "            dataframe.at[index, 'chr'] = 'Chr I'\n",
    "        elif dataframe.at[index, 'chr'] == old_chr_names[1]:\n",
    "            dataframe.at[index, 'chr'] = 'Chr II'\n",
    "        elif dataframe.at[index, 'chr'] == old_chr_names[2]:\n",
    "            dataframe.at[index, 'chr'] = 'Chr III'\n",
    "        elif dataframe.at[index, 'chr'] == old_chr_names[3]:\n",
    "            dataframe.at[index, 'chr'] = 'Chr IV'\n",
    "        elif dataframe.at[index, 'chr'] == old_chr_names[4]:\n",
    "            dataframe.at[index, 'chr'] = 'Chr V'\n",
    "        elif dataframe.at[index, 'chr'] == old_chr_names[5]:\n",
    "            dataframe.at[index, 'chr'] = 'X Chr'\n",
    "            \n",
    "def get_relative_positions(all_df, chr_df):\n",
    "    for index, row in all_df.iterrows():\n",
    "        for cindex, crow in chr_df.iterrows():\n",
    "            if (all_df.at[index, 'Chromosome']) == (chr_df.at[cindex, 'chr']): # if the same chr\n",
    "                all_df.at[index, 'Relative Start Position'] = int(all_df.at[index, 'Start']) / int(chr_df.at[cindex, 'end']) #\n",
    "\n",
    "def make_chr_df_from_fasta(fasta_file: str, species: str):\n",
    "    info = []\n",
    "    counter = 0\n",
    "    for record in SeqIO.parse(fasta_file, \"fasta\"):\n",
    "        \n",
    "        chr_name = chromosomes_roman[kim_counter]\n",
    "        start = 1\n",
    "        end = len(record)\n",
    "        species = species\n",
    "        \n",
    "        info.append([chr_name, start, end, species])\n",
    "        counter = counter + 1\n",
    "    #     print(\"%s %i\" % (record.id, len(record)))\n",
    "\n",
    "    chr_df = pd.DataFrame(info, columns = ['chr', 'start', 'end', 'strain'])\n",
    "    return chr_df\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 4.1.5. <a id='toc4_1_5_'></a>[Define chromosome lengths dataframes](#toc0_)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "id": "7a467570",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2023-01-13T23:59:48.862217Z",
     "start_time": "2023-01-13T23:59:48.851713Z"
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "done\n"
     ]
    }
   ],
   "source": [
    "chr_lengths_file = filebase + 'chr_lengths.txt'\n",
    "\n",
    "\n",
    "N2_chr_df = get_chr_file(chr_lengths_file)[0]\n",
    "CB_chr_df = get_chr_file(chr_lengths_file)[1]\n",
    "\n",
    "# kimcb_fa_file = filebase + '2019_CB4856_genome.fasta'\n",
    "# kimcb_chr_df = make_chr_df_from_fasta(kimcb_fa_file, 'Kim CB4856')\n",
    "\n",
    "# vc_fa_file = filebase + 'N2_vc2010_genome.fasta'\n",
    "# vc_chr_df = make_chr_df_from_fasta(vc_fa_file, 'VC2010')\n",
    "\n",
    "fix_chr_names(N2_chr_df)\n",
    "fix_chr_names(CB_chr_df)\n",
    "# fix_chr_names(kimcb_chr_df)\n",
    "# fix_chr_names(vc_chr_df)\n",
    "\n",
    "print('done')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 4.2. <a id='toc4_2_'></a>[IMPORT TE FILES AND CREATE TE DATAFRAMES](#toc0_)\n",
    "dataframe names:\\\n",
    "N2_all_df, CB_all_df (all tes)\\\n",
    "N2_dna_tes, CB_dna_tes (only DNA (class II) tes)\\\n",
    "N2_retro_tes, CB_retro_tes (only retro (class I) tes)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "7c47e678",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2023-01-14T00:00:26.889369Z",
     "start_time": "2023-01-13T23:59:52.685615Z"
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "done\n"
     ]
    }
   ],
   "source": [
    "N2_file = filebase + 'N2_FinalAnnotations_Transposons.gff3'\n",
    "CB_file = filebase + 'CB_FinalAnnotations_Transposons.gff3'\n",
    "# kimCB_file = '/Users/coraalbers/Library/CloudStorage/OneDrive-UniversityOfOregon/Desktop/labfiles/TEUlt_KimCB4856/kimCB_TEUlt/finalResults/FinalAnnotations_Transposons.gff3'\n",
    "# vc_file = '/Users/coraalbers/Library/CloudStorage/OneDrive-UniversityOfOregon/Desktop/labfiles/TEUlt_VC2010/finalResults/FinalAnnotations_Transposons.gff3'\n",
    "\n",
    "N2_all_list = build_TE_df_from_gff3(N2_file)\n",
    "N2_all_df = N2_all_list[0]\n",
    "N2_data_list = N2_all_list[1]\n",
    "length = ['N2']*len(N2_all_df.Chromosome)\n",
    "N2_all_df['Strain'] = length\n",
    "\n",
    "CB_all_list = build_TE_df_from_gff3(CB_file)\n",
    "CB_all_df = CB_all_list[0]\n",
    "CB_data_list = CB_all_list[1]\n",
    "length2 = ['CB']*len(CB_all_df.Chromosome)\n",
    "CB_all_df['Strain'] = length2\n",
    "\n",
    "# kimCB_all_list = build_TE_df_from_gff3(kimCB_file)\n",
    "# kimCB_all_df = kimCB_all_list[0]\n",
    "# kimCB_data_list = kimCB_all_list[1]\n",
    "# length3 = ['kimCB']*len(kimCB_all_df.Chromosome)\n",
    "# kimCB_all_df['Strain'] = length3\n",
    "\n",
    "# vc_all_list = build_TE_df_from_gff3(vc_file)\n",
    "# vc_all_df = vc_all_list[0]\n",
    "# vc_data_list = vc_all_list[1]\n",
    "# length4 = ['VC2010']*len(vc_all_df.Chromosome)\n",
    "# vc_all_df['Strain'] = length4\n",
    "\n",
    "\n",
    "get_relative_positions(N2_all_df, N2_chr_df)\n",
    "get_relative_positions(CB_all_df, CB_chr_df)\n",
    "# get_relative_positions(kimCB_all_df, kimcb_chr_df)\n",
    "# get_relative_positions(vc_all_df, vc_chr_df)\n",
    "\n",
    "print('done')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# 5. <a id='toc5_'></a>[SIZE FILTERING](#toc0_)\n",
    "- filter out TEs that are outside the reasonable lengths expected of TEs that  size\n",
    "- these parameters are listed in the `range_values` dict in the `create_small_range_dataframe` function and can be changed however needed\n",
    "- alternatively, size filtering can be skipped altogether if wanted\n",
    "## 5.1. <a id='toc5_1_'></a>[FX: `create_small_range_dataframe`](#toc0_)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "id": "9dd7bd90",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2023-01-14T00:00:50.242289Z",
     "start_time": "2023-01-14T00:00:50.231881Z"
    }
   },
   "outputs": [],
   "source": [
    "def create_small_range_dataframe(dataframe):\n",
    "    \n",
    "    range_values = {'CMC':20000, \n",
    "                    'Helitron':20000, \n",
    "                    'Zator':15000, \n",
    "                    'hAT':20000, \n",
    "                    'Sola':7000, \n",
    "                    'Tc1-Mariner':20000,\n",
    "                    'MITE':1000, \n",
    "                    'Novosib':3000,\n",
    "                    'Copia':20000,\n",
    "                    'Gypsy':20000,\n",
    "                    'LINE':10000, \n",
    "                    'ERV':8000, \n",
    "                    'SINE':600}\n",
    "    \n",
    "    filt_counts = {'CMC':0, 'Zator':0, 'hAT':0, 'Gypsy':0, 'Copia':0, 'Sola':0, 'Tc1-Mariner':0,\n",
    "       'MITE':0, 'Helitron':0, 'Novosib':0, 'LINE':0, 'ERV':0, 'SINE':0}\n",
    "    \n",
    "    datalist = dataframe.values.tolist()\n",
    "    \n",
    "    filtered_list = []\n",
    "    elim_list = []\n",
    "    for i in range(len(datalist)):\n",
    "        family_name = datalist[i][6]\n",
    "        if datalist[i][5] <= range_values[family_name]:\n",
    "            filtered_list.append(datalist[i])\n",
    "        else:\n",
    "            elim_list.append(datalist[i])\n",
    "#             filt_counts[family_name] = filt_counts[family_name] + 1\n",
    "            \n",
    "    filtered_dataframe = pd.DataFrame(filtered_list, columns = ['Chromosome', 'Start', 'Stop', 'Score:Probability', 'Strand', 'Size',\n",
    "       'Family', 'Subclass', 'Class', 'Strain', 'Relative Start Position'])\n",
    "    elim_dataframe = pd.DataFrame(elim_list, columns = ['Chromosome', 'Start', 'Stop', 'Score:Probability', 'Strand', 'Size',\n",
    "       'Family', 'Subclass', 'Class', 'Strain', 'Relative Start Position'])\n",
    "    \n",
    "    \n",
    "    original_counts = dataframe['Family'].value_counts().to_dict()\n",
    "    \n",
    "    filtered_counts = filtered_dataframe['Family'].value_counts().to_dict()\n",
    "    elim_counts = elim_dataframe['Family'].value_counts().to_dict()\n",
    "    \n",
    "    strain = filtered_dataframe.at[1, 'Strain']\n",
    "    print(strain, 'fraction elimimated by filter')\n",
    "    for key in original_counts.keys():\n",
    "        filtered = (1- filtered_counts[key]/original_counts[key])\n",
    "        format_filtered = \"{:.5f}\".format(filtered)\n",
    "        print(\"{0:15} {1}\".format(key, format_filtered))\n",
    "\n",
    "#     print(elim_counts)\n",
    "\n",
    "    return filtered_dataframe, elim_dataframe"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 5.2. <a id='toc5_2_'></a>[Generate small range dataframes](#toc0_)\n",
    "- create dataframes from the size filtered TEs\n",
    "- create subsetted dataframes of only DNA and Retro TEs\n",
    "- create a large dataframe with all of the TEs from all the genomes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "id": "9100118d",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2023-01-14T00:00:51.173107Z",
     "start_time": "2023-01-14T00:00:50.866177Z"
    },
    "scrolled": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "N2 fraction elimimated by filter\n",
      "Zator           0.00044\n",
      "hAT             0.00025\n",
      "Tc1-Mariner     0.00107\n",
      "CMC             0.00533\n",
      "Sola            0.00243\n",
      "Gypsy           0.00358\n",
      "Copia           0.00000\n",
      "MITE            0.02817\n",
      "Novosib         0.08000\n",
      "Helitron        0.00000\n",
      "SINE            0.00000\n",
      "ERV             0.00000\n",
      "LINE            0.00000\n",
      "CB fraction elimimated by filter\n",
      "Zator           0.00011\n",
      "hAT             0.00026\n",
      "CMC             0.00248\n",
      "Tc1-Mariner     0.00129\n",
      "Sola            0.00851\n",
      "Gypsy           0.00337\n",
      "MITE            0.02162\n",
      "Copia           0.00000\n",
      "Helitron        0.00000\n",
      "Novosib         0.17647\n",
      "SINE            0.00000\n",
      "LINE            0.00000\n",
      "ERV             0.00000\n",
      "should be 91059 for all 4 genomes: 36479\n"
     ]
    }
   ],
   "source": [
    "N2_all_filter_list = create_small_range_dataframe(N2_all_df)\n",
    "CB_all_filter_list = create_small_range_dataframe(CB_all_df)\n",
    "\n",
    "N2_all_filter_df = N2_all_filter_list[0]\n",
    "CB_all_filter_df = CB_all_filter_list[0]\n",
    "\n",
    "N2_elim_df = N2_all_filter_list[1]\n",
    "CB_elim_df = CB_all_filter_list[1]\n",
    "\n",
    "length = ['N2']*len(N2_all_filter_df.Chromosome)\n",
    "N2_all_filter_df['Strain'] = length\n",
    "\n",
    "length2 = ['CB']*len(CB_all_filter_df.Chromosome)\n",
    "CB_all_filter_df['Strain'] = length2\n",
    "\n",
    "\n",
    "N2_dna_tes = N2_all_filter_df[N2_all_filter_df['Class']=='DNATransposon']\n",
    "N2_retro_tes = N2_all_filter_df[N2_all_filter_df['Class']!='DNATransposon']\n",
    "\n",
    "CB_dna_tes = CB_all_filter_df[CB_all_filter_df['Class']=='DNATransposon']\n",
    "CB_retro_tes = CB_all_filter_df[CB_all_filter_df['Class']!='DNATransposon']\n",
    "\n",
    "# kimCB_all_filter_list = create_small_range_dataframe(kimCB_all_df)\n",
    "# kimCB_all_filter_df = kimCB_all_filter_list[0]\n",
    "# kimCB_elim_df = kimCB_all_filter_list[1]\n",
    "# length3 = ['kimCB']*len(kimCB_all_filter_df.Chromosome)\n",
    "# kimCB_all_filter_df['Strain'] = length3\n",
    "\n",
    "# kimCB_dna_tes = kimCB_all_filter_df[kimCB_all_filter_df['Class']=='DNATransposon']\n",
    "# kimCB_retro_tes = kimCB_all_filter_df[kimCB_all_filter_df['Class']!='DNATransposon']\n",
    "\n",
    "\n",
    "# vc_all_filter_list = create_small_range_dataframe(vc_all_df)\n",
    "# vc_all_filter_df = vc_all_filter_list[0]\n",
    "# vc_elim_df = vc_all_filter_list[1]\n",
    "# length4 = ['VC2010']*len(vc_all_filter_df.Chromosome)\n",
    "# vc_all_filter_df['Strain'] = length4\n",
    "\n",
    "# vc_dna_tes = vc_all_filter_df[vc_all_filter_df['Class']=='DNATransposon']\n",
    "# vc_retro_tes = vc_all_filter_df[vc_all_filter_df['Class']!='DNATransposon']\n",
    "\n",
    "\n",
    "all_filter_df = pd.concat([N2_all_filter_df, CB_all_filter_df]) #, kimCB_all_filter_df, vc_all_filter_df])\n",
    "\n",
    "all_dna_tes = pd.concat([N2_dna_tes, CB_dna_tes]) #, kimCB_dna_tes, vc_dna_tes])\n",
    "all_retro_tes = pd.concat([N2_retro_tes, CB_retro_tes]) #, kimCB_retro_tes, vc_retro_tes])\n",
    "print('should be 91059 for all 4 genomes:', len(all_filter_df))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 5.3. <a id='toc5_3_'></a>[FX: `make_gff3`](#toc0_)\n",
    "- make gff3 files of the size filtered data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "9a276f85",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2023-01-14T00:00:53.668548Z",
     "start_time": "2023-01-14T00:00:53.665236Z"
    }
   },
   "outputs": [],
   "source": [
    "def make_gff3(data_list: list, elim_df, filename: str, to_print: bool = False):\n",
    "    \n",
    "    strain = elim_df.at[1, 'Strain']\n",
    "    print(strain)\n",
    "    print(f'intial:', len(data_list))\n",
    "\n",
    "    pop_list = []\n",
    "\n",
    "    for index, row in elim_df.iterrows():\n",
    "\n",
    "        te_start = int(elim_df.at[index, 'Start'])\n",
    "\n",
    "        for i in range(len(data_list)):\n",
    "\n",
    "            if int(data_list[i][3]) == te_start:\n",
    "\n",
    "                pop_list.append(i)\n",
    "\n",
    "    print('number removed:', len(pop_list))\n",
    "\n",
    "    for i in sorted(pop_list, reverse=True):\n",
    "\n",
    "        data_list.pop(i)      \n",
    "\n",
    "    print('final:', len(data_list))\n",
    "    \n",
    "    if to_print == True:\n",
    "        \n",
    "        with open(f\"{filename}.gff3\", \"w\") as f:\n",
    "            \n",
    "            wr = csv.writer(f, delimiter = '\\t')\n",
    "            \n",
    "            wr.writerows(data_list)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "0662e09f",
   "metadata": {},
   "outputs": [],
   "source": [
    "make_gff3(N2_data_list, N2_elim_df, 'N2_filter', False)\n",
    "# make_gff3(CB_data_list, CB_elim_df, 'CB_filter', False)\n",
    "# make_gff3(kimCB_data_list, kimCB_elim_df, 'kimCB_filter', False)\n",
    "# make_gff3(vc_data_list, vc_elim_df, 'VC2010_filter', False)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "263bdb39",
   "metadata": {},
   "source": [
    "**N2** \\\n",
    "intial: 18419\\\n",
    "number removed: 27\\\n",
    "final: 18392\\\n",
    "**CB**\\\n",
    "intial: 18113\\\n",
    "number removed: 27\\\n",
    "final: 18086\\\n",
    "**kimCB**\\\n",
    "intial: 31227\\\n",
    "number removed: 57\\\n",
    "final: 31170\\\n",
    "**VC2010**\\\n",
    "intial: 23802\\\n",
    "number removed: 68\\\n",
    "final: 23734\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 5.4. <a id='toc5_4_'></a>[STATS: percent of genome that is TE](#toc0_)\n",
    "**based on sum of sizes of all TEs **\n",
    "\n",
    "- N2: 14.68%\n",
    "- CB: 14.30%\n",
    "- VC2010: 15.84%\n",
    "- Kim CB: 16.31%\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "5af62d0d",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2023-01-14T00:00:58.489374Z",
     "start_time": "2023-01-14T00:00:58.483536Z"
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "n2 bp:  14744236\n",
      "n2 frac:  0.14680816341486413\n",
      "\n",
      "cb bp:  14128385\n",
      "cb frac:  0.14296209923841163\n",
      "\n",
      "vc bp:  16168747\n",
      "vc frac:  0.15839557437662422\n",
      "\n",
      "kimcb bp:  16785874\n",
      "kimcb frac:  0.16312666861462582\n"
     ]
    }
   ],
   "source": [
    "n2num = N2_all_filter_df['Size'].sum()\n",
    "n2len = N2_chr_df['end'].sum()\n",
    "print('n2 bp: ', n2num)\n",
    "print('n2 frac: ', n2num / n2len)\n",
    "print()\n",
    "\n",
    "cbnum = CB_all_filter_df['Size'].sum()\n",
    "cblen = CB_chr_df['end'].sum()\n",
    "print('cb bp: ', cbnum)\n",
    "print('cb frac: ', cbnum / cblen)\n",
    "print()\n",
    "\n",
    "vcnum = vc_all_filter_df['Size'].sum()\n",
    "vclen = vc_chr_df['end'].sum()\n",
    "print('vc bp: ', vcnum)\n",
    "print('vc frac: ', vcnum / vclen)\n",
    "print()\n",
    "\n",
    "kimcbnum = kimCB_all_filter_df['Size'].sum()\n",
    "kimcblen = kimcb_chr_df['end'].sum()\n",
    "print('kimcb bp: ', kimcbnum)\n",
    "print('kimcb frac: ', kimcbnum / kimcblen)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# 6. <a id='toc6_'></a>[GRAPHS](#toc0_)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 6.1. <a id='toc6_1_'></a>[graphing variables](#toc0_)\n",
    "- colors, lists and other variables used throughout all of the graphing functions\n",
    "- change these to whatever! just note that the genome names in the `colors` and `strain_dict` dictionaries need to match the genome names you used to label your strains when creating your dataframes\n",
    "\n",
    "<a id='graphing_variables'></a>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 75,
   "id": "8ce62c3d",
   "metadata": {},
   "outputs": [],
   "source": [
    "# N2 #86abb1\n",
    "# CB #df7871\n",
    "# KimCB #b18eaa\n",
    "# VC2010 #2a7f62\n",
    "\n",
    "colors = {'N2':'#86abb1','CB':'#df7871', 'kimCB':'#b18eaa', 'VC2010':'#2a7f62'}\n",
    "# strain_dict = {'N2': [N2_dna_tes, N2_retro_tes], 'CB':[CB_dna_tes, CB_retro_tes], 'VC2010':[vc_dna_tes, vc_retro_tes], 'kimCB':[kimCB_dna_tes, kimCB_retro_tes]}\n",
    "dna_order = ['Zator', 'hAT', 'CMC', 'Tc1-Mariner', 'Sola', 'Novosib', 'Helitron', 'MITE']\n",
    "retro_order = ['Gypsy', 'Copia', 'LINE', 'SINE', 'ERV']\n",
    "te_fam_list = ['CMC', 'hAT', 'Zator', 'Tc1-Mariner', 'Sola', 'Gypsy', 'Copia', 'Novosib', 'Helitron', 'MITE']\n",
    "\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 6.2. <a id='toc6_2_'></a>[FX: `one_x_axis`](#toc0_)\n",
    "- creates a graph with TEs (either all, only DNA or only Retro) from one genome and all chromosomes on one x-axis\n",
    "- x-axis: relative start position\n",
    "- y-axis: count"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 77,
   "id": "3b6d2d52",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2023-01-14T00:01:09.595901Z",
     "start_time": "2023-01-14T00:01:08.832939Z"
    },
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "def one_x_axis(dataframe, te_type: str, title: str, ylim: int, filestr: str, to_save: bool = False):\n",
    "    \"\"\" \n",
    "    dataframe: TE dataframe\n",
    "    te_type: str, one of either 'DNA', 'Retro' or 'all'\n",
    "    title: str, title that will by output on the top of the graph\n",
    "    ylim: int, limit of the y-axis\n",
    "    filestr: str, will be input into {filestr}_global_te_dist_{ylim}yax.pdf to save the graph with a unique name\n",
    "    to_save: bool, whether or not to save the graph as a pdf (if False, graph will only be output into the notebook)\n",
    "    \"\"\"\n",
    "    \n",
    "    sns.set_style('white')\n",
    "    fig, ax = plt.subplots(figsize=(12, 8))\n",
    "\n",
    "    if te_type == 'DNA':\n",
    "        ax = sns.histplot(dataframe[dataframe['Class'] == 'DNATransposon'], \n",
    "                        x = \"Relative Start Position\", \n",
    "                        binwidth=.01, \n",
    "                        stat='count', \n",
    "                        edgecolor = 'white', \n",
    "                        linewidth = 0, \n",
    "                        color = '#86abb1',\n",
    "                        alpha = 1)\n",
    "    elif te_type == 'Retro':\n",
    "        ax = sns.histplot(dataframe[dataframe['Class'] == 'DNATransposon'], \n",
    "                        x = \"Relative Start Position\", \n",
    "                        binwidth=.01, \n",
    "                        stat='count', \n",
    "                        edgecolor = 'white', \n",
    "                        linewidth = 0, \n",
    "                        color = '#86abb1',\n",
    "                        alpha = 1)\n",
    "    elif te_type == 'all':\n",
    "        ax = sns.histplot(dataframe, \n",
    "                        x = \"Relative Start Position\", \n",
    "                        binwidth=.01, \n",
    "                        stat='count', \n",
    "                        edgecolor = 'white', \n",
    "                        linewidth = 0, \n",
    "                        color = '#86abb1',\n",
    "                        alpha = 1)\n",
    "\n",
    "    sns.set(font_scale=1.5)\n",
    "    ax.set_ylabel('Count', fontsize=18)\n",
    "    ax.set_xlabel('Relative Start Position', fontsize=18)\n",
    "    plt.tick_params(axis='both', which='major', labelsize=16)\n",
    "    ax.set(ylim = (0, ylim))\n",
    "    plt.title(title)\n",
    "    print(fig, ax)\n",
    "    \n",
    "    if to_save == True:\n",
    "        \n",
    "        plt.savefig(f'{filestr}_global_te_dist_{ylim}yax.pdf', dpi=300, facecolor = 'w', edgecolor = 'w', format='pdf', bbox_inches='tight')\n",
    "    \n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "id": "f587826e",
   "metadata": {},
   "outputs": [],
   "source": [
    "# one_x_axis(CB_all_filter_df, 'Retro', 'CB retro', 30, 'CB_retro', False)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 6.3. <a id='toc6_3_'></a>[ALL TE HISTOGRAMS](#toc0_)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 6.3.1. <a id='toc6_3_1_'></a>[FX: `make_ultimate_hist_onegen` (DNA and retro for N2, CB, VC2010 or KimCB)](#toc0_)\n",
    "- creates a grid of histograms (using FacetGrid) of DNA and Retro TEs on one genome\n",
    "- columns: chromosomes, rows: DNA and Retro\n",
    "- x-axis: TE start position (can easily be changed within the function to graph by Relative Start Position if needed)\n",
    "- y-axis: Count"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 46,
   "id": "9185c8e6",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2022-08-04T16:25:12.671256Z",
     "start_time": "2022-08-04T16:25:02.950869Z"
    }
   },
   "outputs": [],
   "source": [
    "def make_ultimate_hist_onegen(dataframe, genome: str, to_save: bool = False):\n",
    "    \"\"\" \n",
    "    dataframe: TE dataframe\n",
    "    genome: str, name of the genome that will be inserted into the title\n",
    "    to_save: bool, whether or not to save the graph as a pdf (if False, graph will only be output into the notebook)\n",
    "\n",
    "    \"\"\"\n",
    "\n",
    "    sns.set(font_scale=1)\n",
    "    sns.set_style('white')\n",
    "\n",
    "    color = colors[genome]\n",
    "\n",
    "    ultimate_hist = sns.FacetGrid(dataframe, col=\"Chromosome\", row=\"Class\", sharex='col', sharey='row', margin_titles=True)# hue_kws={'color':['mediumblue', 'deepskyblue','mediumblue', 'deepskyblue','mediumblue', 'deepskyblue']})\n",
    "    ultimate_hist.map(sns.histplot, \"Start\", binwidth=100000, stat='count', color = color, edgecolor = color)\n",
    "    ultimate_hist.fig.subplots_adjust(top=.85) # move subplots down so that title doesn't overlap with subtitles\n",
    "    ultimate_hist.fig.suptitle(f'Distribution of DNA TEs and Retro-TEs on the {genome} genome')\n",
    "    # ultimate_hist.set(ylim = (0, 250))\n",
    "\n",
    "    if to_save == True: \n",
    "        \n",
    "        plt.savefig(f'TEUlt_{genome}_ultimate_hist_100kb_LTRRet_filter.pdf', dpi=300, facecolor = 'w', edgecolor = 'w', format='pdf', transparent=True)      \n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 48,
   "id": "8141719c",
   "metadata": {},
   "outputs": [],
   "source": [
    "# make_ultimate_hist_onegen(N2_all_filter_df, 'N2', to_save = False)        "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 6.3.2. <a id='toc6_3_2_'></a>[FX: `make_ultimate_hist_allgen` (all genomes for DNA or retro)](#toc0_)\n",
    "- creates a grid of histograms (using FacetGrid) of TEs on all genomes in the all_filter_df dataframe and its subsets\n",
    "- columns: chromosomes, rows: genome\n",
    "- x-axis: TE start position, this can easily be changed within the function to graph by Relative Start Position if needed, the x-axis ticks are manually input as a list within the function and can be changed based on the length of the genomes you are analyzing\n",
    "- y-axis: Count"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 51,
   "id": "e78a0662",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2022-11-16T21:17:08.752635Z",
     "start_time": "2022-11-16T21:17:02.325990Z"
    }
   },
   "outputs": [],
   "source": [
    "def make_ultimate_hist_allgen(dataframe, title: str, filestr: str, binsize: int, to_save: bool = False):\n",
    "    \"\"\" \n",
    "    dataframe: TE dataframe\n",
    "    title: str, name of the type of TEs (ex. DNA, Retro or All) that will be inserted into the title\n",
    "    filestr: str, will be input into TEUlt_{filestr}_allgenomes.pdf to save the graph with a unique name\n",
    "    binsize: int, size of the histogram bins \n",
    "    to_save: bool, whether or not to save the graph as a pdf (if False, graph will only be output into the notebook)\n",
    "\n",
    "    \"\"\"\n",
    "\n",
    "    sns.set(font_scale=1)\n",
    "    sns.set_style('white')\n",
    "    \n",
    "    ultimate_hist = sns.FacetGrid(dataframe, col=\"Chromosome\", row=\"Strain\", sharex='col', sharey=True, margin_titles=True, hue='Strain', palette=colors)# hue_kws={'color':['mediumblue', 'deepskyblue','mediumblue', 'deepskyblue','mediumblue', 'deepskyblue']})\n",
    "\n",
    "    ultimate_hist.map(sns.histplot, \"Start\", binwidth=binsize, stat='count', edgecolor = 'white', linewidth = 0, alpha=1)\n",
    "\n",
    "    ultimate_hist.fig.subplots_adjust(top=.85) # move subplots down so that title doesn't overlap with subtitles\n",
    "\n",
    "    ultimate_hist.fig.suptitle(f'Distribution of {title} TEs by genome')\n",
    "\n",
    "    ultimate_hist.set(xticks=[0, 5000000, 10000000, 15000000, 20000000])\n",
    "\n",
    "    for ax in ultimate_hist.axes.flatten():\n",
    "        ax.tick_params(labelbottom=True, bottom=True, left=True, direction='out', length=6, width=2)\n",
    "        \n",
    "    if to_save == True:\n",
    "\n",
    "        plt.savefig(f'TEUlt_{filestr}_allgenomes.pdf', dpi=300, facecolor = 'w', edgecolor = 'w', format='pdf')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 53,
   "id": "2ed77989",
   "metadata": {},
   "outputs": [],
   "source": [
    "# make_ultimate_hist_allgen(all_dna_tes, 'DNA', 'dna_500kb', 500000, to_save = False)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 6.4. <a id='toc6_4_'></a>[SPECIFIC TE FAMILY HISTOGRAMS](#toc0_)\n",
    "### 6.4.1. <a id='toc6_4_1_'></a>[FX: `make_spec_family_ult_hist`](#toc0_)\n",
    "- plot distribution of a specific TE family for one genome\n",
    "- x-axis: Start Position (can be changed to Relative Start Position within the function)\n",
    "- y-axis: Count"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 54,
   "id": "61ce921d",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2022-11-28T22:05:52.696221Z",
     "start_time": "2022-11-28T22:05:52.690174Z"
    }
   },
   "outputs": [],
   "source": [
    "def make_spec_family_ult_hist(all_df, genome: str, spec_family: str, binsize: int, to_save: bool = False):\n",
    "    \"\"\" \n",
    "    all_df: dataframe with all TEs from all genomes\n",
    "    genome: str, name of the genome that will be inserted into the title\n",
    "    spec_family: str, name of the specific TE family that will be inserted into the title \n",
    "    binsize: int, size of the histogram bins \n",
    "    to_save: bool, whether or not to save the graph as a pdf (if False, graph will only be output into the notebook)\n",
    "    \"\"\"\n",
    "    \n",
    "    sns.set_style('white')\n",
    "    \n",
    "    color = colors[genome]\n",
    "    \n",
    "    spec_family_ultimate_hist = sns.FacetGrid(all_df[all_df['Family'] == spec_family], col=\"Chromosome\", sharex='col', sharey='row', margin_titles=True)# hue_kws={'color':['mediumblue', 'deepskyblue','mediumblue', 'deepskyblue','mediumblue', 'deepskyblue']})\n",
    "    spec_family_ultimate_hist.map(sns.histplot, \"Start\", binwidth=binsize, stat='count', color = color, edgecolor= color)\n",
    "    spec_family_ultimate_hist.fig.subplots_adjust(top=.8) # move subplots down so that title doesn't overlap with subtitles\n",
    "    spec_family_ultimate_hist.fig.suptitle(f'Distribution of {spec_family} TEs on the {genome} genome')\n",
    "    \n",
    "    binfilestr = str(int(binsize/1000)) +'kb'\n",
    "    \n",
    "    if to_save == True:\n",
    "        \n",
    "        plt.savefig(f'TEUlt_{genome}_{spec_family}_{binfilestr}hist.pdf', dpi=300, facecolor = 'w', edgecolor = 'w', format='pdf')\n",
    "        \n",
    "    return spec_family_ultimate_hist"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 60,
   "id": "7b7400f4",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2022-07-11T20:50:08.140876Z",
     "start_time": "2022-07-11T20:50:08.138382Z"
    }
   },
   "outputs": [],
   "source": [
    "# make_spec_family_ult_hist(N2_all_df, 'N2', 'CMC', 100000, False)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "below: runs the `make_spec_family_ult_hist` function on all TE families in the `te_fam_list` defined in the [graphing variables](#61-graphing-variables) cell"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 59,
   "id": "f6180656",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2022-07-11T20:50:19.496209Z",
     "start_time": "2022-07-11T20:50:08.707742Z"
    }
   },
   "outputs": [],
   "source": [
    "# genome = 'N2'\n",
    "# color = '#4c72b0'\n",
    "# df = N2_all_filter_df\n",
    "\n",
    "# for i in range(len(te_fam_list)):\n",
    "#     spec_family = te_fam_list[i]\n",
    "#     make_spec_family_ult_hist(df, genome, spec_family, 100000, False)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 6.4.2. <a id='toc6_4_2_'></a>[FX: `make_spec_family_ult_hist_allgenomes`](#toc0_)\n",
    "- plot distribution of a specific transposon family by chromosome for both all genomes\n",
    "- columns: chromosomes, rows: genome\n",
    "- x-axis: relative start position\n",
    "- y-axis: count"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 66,
   "id": "e1be5ac7",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2022-08-08T05:42:09.645365Z",
     "start_time": "2022-08-08T05:42:09.638190Z"
    }
   },
   "outputs": [],
   "source": [
    "def make_spec_family_ult_hist_allgenomes(all_df, spec_family: str, binsize: int, to_save: bool = False):\n",
    "    \"\"\" \n",
    "    all_df: dataframe with all TEs from all genomes\n",
    "    spec_family: str, name of the specific TE family that will be inserted into the title \n",
    "    binsize: int, size of the histogram bins \n",
    "    to_save: bool, whether or not to save the graph as a pdf (if False, graph will only be output into the notebook)\n",
    "    \"\"\"\n",
    "\n",
    "    sns.set_style('white')\n",
    "\n",
    "    spec_family_ultimate_hist_allgen = sns.FacetGrid(all_df[all_df['Family'] == spec_family], \n",
    "                                                      col=\"Chromosome\", \n",
    "                                                      row = 'Strain', \n",
    "                                                      sharex='col', \n",
    "                                                      sharey=True, \n",
    "                                                      margin_titles=True, \n",
    "                                                      hue = 'Strain', \n",
    "                                                      palette = colors)  # hue_kws={'color':['mediumblue', 'deepskyblue','mediumblue', 'deepskyblue','mediumblue', 'deepskyblue']})\n",
    "    \n",
    "    spec_family_ultimate_hist_allgen.map(sns.histplot, \"Relative Start Position\", binwidth=binsize, stat='count', edgecolor = 'white', linewidth = 0, alpha=1)\n",
    "    spec_family_ultimate_hist_allgen.fig.subplots_adjust(top=.8) # move subplots down so that title doesn't overlap with subtitles\n",
    "    spec_family_ultimate_hist_allgen.fig.suptitle(f'Distributions of {spec_family} TEs')\n",
    "    \n",
    "    for ax in spec_family_ultimate_hist_allgen.axes.flatten():\n",
    "        ax.tick_params(labelbottom=True)\n",
    "    \n",
    "    binfilestr = str(binsize)[2:]\n",
    "    \n",
    "    if to_save == True:\n",
    "        \n",
    "        plt.savefig(f'{spec_family}_{binfilestr}_allgenomes.pdf', dpi=300, facecolor = 'w', edgecolor = 'w', format='pdf')\n",
    "    \n",
    "    return spec_family_ultimate_hist_allgen\n",
    "  "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 68,
   "id": "27ebee54",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2022-08-08T05:48:09.385909Z",
     "start_time": "2022-08-08T05:48:04.001495Z"
    },
    "scrolled": false
   },
   "outputs": [],
   "source": [
    "# make_spec_family_ult_hist_allgenomes(all_dna_tes, 'Zator', .01, False)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "below: runs the `make_spec_family_ult_hist_allgenomes` function on all TE families in the `dna_te_fam_list` or `retro_te_fam_list`"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 69,
   "id": "7a393368",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2022-07-28T23:05:29.582231Z",
     "start_time": "2022-07-28T23:05:20.196737Z"
    },
    "scrolled": false
   },
   "outputs": [],
   "source": [
    "# dna_te_fam_list = ['CMC', 'hAT', 'Zator', 'Tc1-Mariner', 'Sola', 'Novosib']#, 'Helitron']#, 'MITEs']\n",
    "# retro_te_fam_list = ['Gypsy', 'Copia']\n",
    "\n",
    "# df = all_dna_tes\n",
    "# te_type = 'dna'\n",
    "# data_list = dna_te_fam_list \n",
    "\n",
    "# for i in range(len(data_list)):\n",
    "#     spec_family = data_list[i]\n",
    "#     make_spec_family_ult_hist_allgenomes(df, spec_family, .01, False)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 6.4.3. <a id='toc6_4_3_'></a>[FX: `track_down_hist_peaks`](#toc0_)\n",
    "- basic function to look at histogram peaks in the N2 and CB specific family data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 91,
   "id": "19512a58",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2022-08-02T23:31:39.061788Z",
     "start_time": "2022-08-02T23:31:39.052342Z"
    }
   },
   "outputs": [],
   "source": [
    "def track_down_hist_peaks(N2_df, CB_df, spec_family, chromosome):\n",
    "    sns.set_style('white')\n",
    "    \n",
    "    N2data = N2_df.loc[(N2_df['Family'] == spec_family) & (N2_df['Chromosome'] == chromosome)]['Start'].tolist()\n",
    "    CBdata = CB_df.loc[(CB_df['Family'] == spec_family) & (CB_df['Chromosome'] == chromosome)]['Start'].tolist()\n",
    "\n",
    "    \n",
    "    values, bins, patches = plt.hist(N2data, bins=200, color = 'black')\n",
    "    plt.title('N2')\n",
    "    plt.show()\n",
    "    \n",
    "    values1, bins1, patches1 = plt.hist(CBdata, bins=200, color = 'black')\n",
    "    plt.title('CB')\n",
    "    plt.show()\n",
    "\n",
    "    return [values, bins, patches, values1, bins1, patches1]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 92,
   "id": "7b42b2f6",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2022-08-03T00:10:39.951109Z",
     "start_time": "2022-08-03T00:10:39.325712Z"
    }
   },
   "outputs": [
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAyAAAAMgCAYAAADbcAZoAAAAOXRFWHRTb2Z0d2FyZQBNYXRwbG90bGliIHZlcnNpb24zLjUuMSwgaHR0cHM6Ly9tYXRwbG90bGliLm9yZy/YYfK9AAAACXBIWXMAAB7CAAAewgFu0HU+AABNoElEQVR4nO3de7xVZZ0/8M9B7gEiFxH1iEylohhlmjmmFKiRqWkGRmhkIahZqDXqTBZmOc2MOM7IJCqUl1HCMZ2kSATFMUZTKS0F1MpEvJAiF8G4KZzfH/zYc7gdOIdz1mGf836/Xr5ez97redb6rr3X2eyPa+31VFRVVVUFAACgAC0auwAAAKD5EEAAAIDCCCAAAEBhBBAAAKAwAggAAFAYAQQAACiMAAIAABRGAAEAAAojgAAAAIURQAAAgMIIIAAAQGEEEAAAoDACCAAAUBgBBAAAKIwAAgAAFEYAAQAACiOAAAAAhRFAAACAwgggAABAYQQQAACgMAIIAABQGAEEAAAojAACAAAURgABAAAKI4AAAACFEUAAAIDCCCAAAEBhBBAAAKAwAggAAFAYAQQAACiMAAIAABRGAAEAAAojgAAAAIURQAAAgMIIIAAAQGEEEAAAoDACCAAAUBgBBAAAKIwAAgAAFEYAAQAACiOAAAAAhRFAAACAwgggAABAYQQQAACgMAIIAABQGAEEAAAojAACAAAURgABAAAKI4AAAACFEUAAAIDCCCAAAEBhBBAAAKAwAggAAFAYAQQAACiMAAIAABRGAAEAAAojgAAAAIURQADYwhVXXJGKiopUVFSkU6dOWbBgQa3Gf/zjHy+Nf+CBB7bZ74knnsi5556bvn37Zvfdd0/r1q3To0ePDBgwINdcc02WL1++s7sCwC5GAAGgRitWrMiIESPqdZ1//etf88UvfjFHHnlkbrzxxsydOzfLly/PO++8kzfeeCMPPfRQvvnNb+a9731vpk+fXq/bBqBxtWzsAgDY9c2YMSMTJkzIOeecs9PrWrduXU4//fTcf//9pec+/vGP52//9m/znve8J3/+85/zs5/9LIsXL86bb76Zk046Kffdd18GDhy409sGoPEJIADskG9+85sZNGhQKisrd2o9N910Uyl8dOnSJffcc0/69++/SZ9rr702w4cPz3//93/nnXfeyfDhw/PHP/4x7dq126ltA9D4XIIFwA5Zvnx5vZwBueaaa0rtW2+9dYvwkSQdO3bM5MmTc+ihhyZJXn311dx55507vW0AGp8AAkCNjjnmmFRUVCRJ7r///vzoRz+q87rmzp2bF154IUlywAEH5KSTTtpm39atW+e8884rPX7wwQfrvF0Adh0CCAA1OvbYYzcJAt/4xjfyyiuv1Gldc+bMKbWPOOKI7fZ/73vfW2ovXLiwTtsEYNcigACwXf/8z/+c/fffP0ny1ltvZeTIkXVaz+mnn54FCxbksccey6WXXrrd/q+99lqp3aFDhzptE4BdiwACwHZ16NAhP/rRj0qXYt133325+eaba72eli1bprKyMkceeWTp9x01ufvuu0vtQw45pNbbA2DXI4AAsEMGDBiQUaNGlR5ffPHFefXVVxtse4888kimTp1aenz66ac32LYAKI4AAsAOu/rqq9OrV68kybJlyzYJJPXpzTffzFlnnZWqqqokyUknnZTDDjusQbYFQLEEEAB22MZLsTaaOnVqbr311nrdxooVK3LyySfnxRdfTJJ07tw5//7v/16v2wCg8QggANTKwIEDN/kR+kUXXVRvd6hatmxZBg0alMceeyxJ0qJFi9x22235m7/5m3pZPwCNTwABoNbGjh2b/fbbL0mydOnSerkU67XXXsuxxx6bRx99NMmG8DFx4sScfPLJO71uAHYdAggAtdaxY8dMnDix9PjnP/95/vM//7PO63vqqafykY98JM8880ySDXfLuuWWW3L22WfvdK0A7FoEEADq5Pjjj88555xTejx69Oj85S9/qfV6pkyZkmOOOaZ0R6127drl7rvvzllnnVVvtQKw6xBAAKizsWPHprKyMsmGS7HOPffcWo3/j//4j5x22mn561//miTp1q1bZs6cmVNOOaXeawVg1yCAAFBnnTp12uRSrHvvvTd33HHHDo39x3/8x3zta1/L+vXrkyTvf//78+tf/zof/ehHG6RWAHYNAggAO+WEE07IV77yldLjr3/963njjTdqHHPdddflW9/6VunxkUcemUcffTTve9/7GqxOAHYNAggAO+1f//Vfs++++yZJlixZkmeffXabff/3f/83F198cenxgAED8uCDD6Zbt24NXicAjU8AAWCnderUKRMmTNhuv7Vr1+ass87KunXrkiT9+vXLvffem/e85z0NXSIAu4iWjV0AAE3DoEGDcvbZZ+fmm2/eZp+bb7458+fPLz0+4ogjcsMNN+zQ+vfdd998/vOf39kyAWhkAggA9ebaa6/N9OnTS7fU3dwtt9yyyePqP2DfnqOPPloAAWgCXIIFQL3Zfffdc9NNN21z+caJBgFoviqqqqqqGrsIAACgeXAGBAAAKIwAAgAAFEYAAQAACiOAAAAAhRFAAACAwgggAABAYQQQAACgMAIIAABQGAEEAAAojAACAAAURgABAAAK07KxC2CDhQsXZubMmaXH++23X9q1a9eIFQEAUN2qVauyYMGC0uMBAwakZ8+ejVhReRJAdhEzZ87MlVde2dhlAABQC8OGDWvsEsqOS7AAAIDCCCAAAEBhXIK1i6isrNzk8Xe+850ceOCBjVQNAACbe/755ze5ZH7z72/sGAFkF9G+fftNHh944IE5/PDDG6kaAAC2Z/Pvb+wYl2ABAACFEUAAAIDCCCAAAEBhBBAAAKAwAggAAFAYAQQAACiMAAIAABRGAAEAAAojgAAAAIURQAAAgMIIIAAAQGEEEAAAoDACCAAAUBgBBAAAKIwAAgAAFEYAAQAACiOAAAAAhRFAAACAwgggAABAYQQQAACgMAIIAABQGAEEAAAojAACAAAURgABAAAKI4AAAACFEUAAAIDCCCAADWDdunVZt25dY5cBALscAQSgASxcuDALFy5s7DIAYJcjgAAAAIURQAAAgMIIIAAAQGEEEAAAoDACCAAAUBgBBAAAKIwAAgAAFEYAAQAACiOAAAAAhRFAAACAwgggAABAYQQQAACgMAIIAABQGAEEAAAojAACAAAURgABAAAKI4AAAACFEUAAAIDCCCAAAEBhBBAAAKAwAggAAFAYAQQAACiMAAIAABSmSQSQ5557Ll//+tdz0EEHpUOHDunQoUP69euXb33rW1m0aFGNY995551cf/31Ofroo7P77runXbt2ef/735+vfvWref755wvaAwAAaB5aNnYBO+vaa6/NZZddlrVr127y/NNPP52nn346EydOzL333puPfvSjW4xdvHhxPvWpT2X27NmbPP+nP/0pf/rTn3LzzTfnxhtvzFlnndWg+wAAAM1FWQeQcePG5eKLL06StG/fPl/5ylfykY98JCtXrszkyZPz0EMP5Y033siJJ56YefPmZa+99iqNXbduXU477bRS+DjyyCMzfPjwdOrUKf/7v/+bH/3oR1m1alW+/OUvp7KyMh//+McbYxcBAKBJKdsAMn/+/Fx66aVJku7du2fmzJnp27dvafnIkSMzevToXHfddVm6dGmuuuqqjBs3rrT8xz/+cWbNmpUkOfPMM3PrrbemRYsNV6QNGzYsQ4cOzSc/+cmsXr06559/fubMmVNaDgAA1E3ZfqP+7ne/m1WrViVJ/uu//muT8LHRv/zLv6R79+6lPtVdc801SZI99tgjP/zhD7cIF8cee2y+853vJEmeffbZ/OxnP6vvXQAAgGanLAPImjVrcs899yRJTjnllG1eHtWmTZtcccUVGT16dL72ta9lzZo1SZLf/e53pR+YDxs2LJ06ddrq+PPOOy+77bZbkuSuu+6q570AAIDmpywvwZo5c2aWL1+eJBk+fHiNfc8///wtnnv44YdL7YEDB25zbOfOnXPYYYdl9uzZuf/+++tYLQAAsFFZngH53e9+V2pXv7vVm2++mf/93//NjBkzMn/+/G2OnzNnTqm9tUu3qjv44IOTJEuXLs2CBQvqVjAAAJCkTAPI3Llzk2y4xGrvvffOCy+8kFNPPTU9e/bMMccckxNOOCG9e/fORz7ykTz66KNbjN8YTlq0aJH99tuvxm1VVlaW2i+99FL97QQAADRDZRlAXnnllSQbfkD+q1/9Kh/84Adz77335t13392k3+zZs9O/f/9Mnjx5k+c3Tk7YsWPHtG7dusZtdenSpdRevHhxfZQPAADNVlkGkBUrViRJVq5cmc9+9rN5++2385WvfCVz5szJ6tWr88ILL+TSSy9NixYt8u677+ZLX/rSJpdtrVy5MknStm3b7W6rep+N4wAAgLopywDy9ttvJ0mWL1+exYsX58orr8zEiRNzyCGHpE2bNvmbv/mb/NM//VOuv/76JBvumnXJJZeUxm88U9KmTZvtbqt6n83PsAAAALVTlgGkqqqq1D700ENz+eWXb7XfqFGjcuSRRyZJZsyYkZdffjlJ0q5duyTJ2rVrt7utjbfuTbLdy7UAAICalWUAec973lNqf/7zn09FRcU2+w4ePLjUfuSRR5IkHTp0SJKsXr16u9uq3mdjcAEAAOqmLANI9YkDDzrooBr7HnjggaX2q6++mmTD/B7Jhku4tndZ1ZIlS0rtjbOqAwAAdVOWAaR379473Lf6bzjWr1+fJDnggANKjzeGkm3ZeNlWkvTq1as2ZQIAAJspywBy6KGHlto1TTiYJK+//nqpvc8++yT5v8kFk/+bU2RbNi7fY489SuMBAIC6KcsA0r9//1L7vvvuq7Hvr3/961J7Y3CpPv6hhx7a5thly5blqaee2mIMAABQN2UZQA4//PC8733vS5I8+OCDmT179lb7LV68OD/5yU+SJH369CkFkIMPPrj025FbbrmldFvfzf3whz/MunXrkiRDhgyp130AAIDmqCwDSJJ85zvfSbLhlrxf+MIXsmDBgk2Wr1mzJmeddVaWLl2aJLnooos2WX7xxRcnSd58880MHz58i1vy/upXv8r3v//9JMn++++f008/vUH2AwAAmpOWjV1AXZ111ln52c9+lnvuuSd/+tOf8oEPfCAjRozIYYcdljfeeCM33nhjnnvuuSQbLp8aMWLEJuO/8pWvZOLEiXniiSdyzz335IgjjsiIESPStWvXPPLII5kwYULeeeedVFRUZPz48eYAAQCAelC2ASRJJk+enBEjRuS2227LW2+9lWuuuWaLPieccEJ++tOfbjFXSIsWLfKLX/wigwYNypNPPpmnn346X//61zfp06pVq1x//fUZNGhQg+4HAAA0F2V7CVayISDceuutmTlzZoYOHZrKysq0adMme++9d44//vj89Kc/zX333ZeOHTtudXz37t3z+OOP54c//GGOPvro7LHHHmnVqlX222+/DB8+PL/97W+3OHMCAADUXVmfAdnoE5/4RD7xiU/UaWzLli1z/vnn5/zzz6/nqgAAgM2V9RkQAACgvAggAABAYQQQAACgMAIIAABQGAEEAAAojAACAAAURgABAAAKI4AAAACFEUAAAIDCCCAAAEBhBBAAAKAwAggAAFAYAQQAACiMAAIAABRGAAEAAAojgAAAAIURQAAAgMIIIAAAQGEEEAAAoDACCAAAUBgBBAAAKIwAAgAAFEYAAQAACiOAAAAAhRFAAACAwgggAABAYQQQAACgMAIIAABQGAEEAAAojAACAAAURgABAAAKI4AAAACFEUAAAIDCCCAAAEBhBBAAAKAwAggAAFAYAQQAACiMAAIAABRGAAEAAAojgAAAAIURQAAAgMIIIAAAQGEEEAAAoDACCAAAUBgBBAAAKIwAAgAAFEYAAQAACiOAAAAAhRFAAACAwgggAABAYQQQAACgMAIIAABQGAEEAAAojAACAAAURgABAAAKI4AAAACFEUAAAIDCCCAAAEBhBBAAAKAwAggAAFAYAQQAACiMAAIAABRGAAEAAAojgAAAAIURQAAAgMIIIAAAQGEEEAAAoDBlHUAefvjhVFRU7NB/gwYN2uo6qqqqMmnSpBx33HHp2rVr2rZtm969e2f48OGZPXt2wXsEAABNW8vGLmBnPP300zs1ftWqVRk8eHCmTp26yfPz58/P/Pnzc8cdd+Sqq67KpZdeulPbAQAANmgSAaRjx4657bbbauzbo0ePLZ4bMWJEKXz06dMnI0eOTI8ePfLkk0/mhhtuyNtvv53LLrss++yzT84888z63wEAAGhmmkQA6du3b0499dRajZ0xY0YmTZqUJBkwYECmTp2atm3bJkmGDh2aL3/5yznmmGOyePHiXHjhhfnMZz6Tjh071mv9AADQ3JTtb0DWr1+fuXPnJtkQQGpr7NixSZKWLVtm4sSJpfCxUZ8+fTJu3LgkyeLFizNhwoSdrBgAACjbAPLCCy/kr3/9a5Lk0EMPrdXYJUuW5IEHHkiSDBo0KL17995qvzPOOCN77rlnkuSuu+7aiWoBAICkjANI9R+g1zaAzJo1K+vXr0+SDBw4cJv9WrRokf79+ydJHn/88SxdurQOlQIAABuVbQB55plnSu2Nl2D9+c9/zvTp0zNz5swsWLBgm2PnzJmzxdhtOfjgg5NsuF1v9W0CAAC1V7YBZOMZkB49euSXv/xlDj744Lz3ve/NJz/5yQwcODC9evXKhz/84UybNm2LsfPnzy+1999//xq3U1lZWWq/9NJL9VI7AAA0V2UfQF5//fUMHz48zz777BZ9nnzyyXzqU5/KlVdeucnzixYtKrW7detW43a6dOlSai9evHhnSgYAgGavLAPIX//61/z5z38uPX7f+96X//zP/8zChQuzatWqPP3007ngggtSUVGRJBkzZkxuueWWUv+VK1eW2pvf/Wpz1ZdXHwcAANReWQaQF154oRQMjjrqqDz55JM588wzs9dee6Vt27Y59NBDM27cuNx8882lMd/85jfz9ttvJ0nefffd0vNt2rSpcVvVl1cfBwAA1F5ZBpAPfOAD+etf/5pXX30199133zYnCBw+fHhOOeWUJBsun7rnnnuSJO3atSv1Wbt2bY3bWrNmTandunXrnS0dAACatbIMIElSUVGRvffeO7vvvnuN/c4666xSe9asWUmSDh06lJ5bvXp1jeOrL68eXAAAgNor2wCyow466KBS+7XXXkuSdO7cufTckiVLahxffXn37t3rtzgAAGhmmnwAadWq1RbPHXDAAaV2TfOFJMnLL79cavfq1av+CgMAgGaoLAPItGnT8k//9E+5+OKLt3tr3FdffbXU7tmzZ5L/m1wwSebOnVvj+I3LKyoqcsghh9S1ZAAAIGUaQO6+++78/d//fa699tr86le/qrHvAw88UGp/9KMfTbLhzlkbz4w89NBD2xy7bt26PPzww0mSfv36bXLpFgAAUHtlGUAGDBhQat94443b7Ldo0aLcdNNNSZL27dvn9NNPT7LhNyADBw5MkkyZMmWbl2FNnjy5NGnhkCFD6qV2AABozsoygJx66qnZd999kyT3339/rrnmmi36LFu2LJ/97GdLl2h94xvfyB577FFaftFFFyXZcBveoUOHZsWKFZuMnzdvXkaPHp0k6dixY84555wG2RcAAGhOWjZ2AXXRrl273HjjjTnllFOybt26fPOb38z999+fz33uc9l9990zd+7cTJgwIX/5y1+SJP3798/ll1++yTpOOOGEnH766bn77rvz6KOPpl+/fjnvvPNSWVmZp556KuPHjy+FkquvvjrdunUrfD8BAKCpKcsAkiQnnnhifvKTn+TLX/5y3n777cyYMSMzZszYot8pp5ySO+64Y6uTCN52221ZsWJFpk+fnhdffDGXXHLJJssrKioyZsyYjBo1qsH2AwAAmpOyvARro8GDB+cPf/hD/uEf/iEf+tCH0rFjx7Rp0yb77bdfhgwZkqlTp+bee+/dZOLB6tq3b59p06bl9ttvz/HHH5/u3bunVatW6dmzZwYPHpxZs2ZlzJgxBe8VAAA0XWV7BmSjnj175qqrrspVV11Vp/EVFRUZNmxYhg0bVs+VAQAAmyvrMyAAAEB5EUAAAIDCCCAAAEBhBBAAAKAwAggAAFAYAQQAACiMAAIAABRGAAEAAAojgAAAAIURQAAAgMIIIAAAQGEEEAAAoDACCAAAUBgBBAAAKIwAAgAAFEYAAQAACiOAAAAAhRFAAACAwgggAABAYQQQAACgMAIIAABQGAEEAAAojAACAAAURgABAAAKI4AAAACFEUAAAIDCCCAAAEBhBBAAAKAwAggAAFAYAQQAACiMAAIAABRGAAEAAAojgAAAAIURQAAAgMIIIAAAQGEEEAAAoDACCAAAUBgBBAAAKIwAAgAAFEYAAQAACiOAAAAAhRFAAACAwgggAABAYQQQAACgMAIIAABQGAEEAAAojAACAAAURgABAAAKI4AAAACFEUAAAIDCCCAAAEBhBBAAAKAwAggAAFAYAQQAACiMAAIAABRGAAEAAAojgAAAAIURQAAAgMIIIAAAQGEEEAAAoDACCAAAUBgBBAAAKIwAAgAAFEYAAQAACiOAAAAAhRFAAACAwgggAABAYQQQAACgME0ygCxbtiz77LNPKioq8qUvfanGvlVVVZk0aVKOO+64dO3aNW3btk3v3r0zfPjwzJ49u5iCAQCgmWjZ2AU0hAsvvDCvvfbadvutWrUqgwcPztSpUzd5fv78+Zk/f37uuOOOXHXVVbn00ksbqlQAAGhWmlwAmTp1am699dYd6jtixIhS+OjTp09GjhyZHj165Mknn8wNN9yQt99+O5dddln22WefnHnmmQ1ZNgAANAtNKoAsW7YsI0eO3KG+M2bMyKRJk5IkAwYMyNSpU9O2bdskydChQ/PlL385xxxzTBYvXpwLL7wwn/nMZ9KxY8cGqx0AAJqDJvUbkI2XXnXu3Hm7fceOHZskadmyZSZOnFgKHxv16dMn48aNS5IsXrw4EyZMqPd6AQCguWkyAaT6pVcbw8W2LFmyJA888ECSZNCgQendu/dW+51xxhnZc889kyR33XVXPVYLAADNU5MIINUvvRo5cmQGDhxYY/9Zs2Zl/fr1SVJj3xYtWqR///5JkscffzxLly6tp4oBAKB5ahIBZOOlV5WVlbn66qu323/OnDmldt++fWvse/DBByfZcLveZ555ZucKBQCAZq7sA0j1S69uvPHGdOrUabtj5s+fX2rvv//+NfatrKwstV966aU61QgAAGxQ1gGk+qVXX/ziF/OpT31qh8YtWrSo1O7WrVuNfbt06VJqL168uA5VAgAAG5V1ANl46dVee+2Vf/u3f9vhcStXriy1N7/71eaqL68+DgAAqL2yDSDVL70aP3589thjjx0e++6775babdq0qbFv9eXVxwEAALVXlgGk+qVXZ5xxRk499dRajW/Xrl2pvXbt2hr7rlmzptRu3bp1rbYDAABsqiwDyMZLr7p161aaLLA2OnToUGqvXr26xr7Vl1cPLgAAQO2VXQCpfunVuHHj0r1791qvo/pM6UuWLKmxb/XlddkWAADwf8ougFSfkXzo0KGpqKjY4r/qM5vfeuutpeevuOKKJMkBBxxQWr5gwYIat/fyyy+X2r169aqnvQAAgOap7AJIfdg4uWCSzJ07t8a+G5dXVFTkkEMOadC6AACgqWvZ2AXU1te//vXt/uj8jTfeyKhRo5Ikn/jEJ/L1r389SXLQQQclSY466qi0atUq77zzTh566KGcf/75W13PunXr8vDDDydJ+vXrt8mlWwAAQO2VXQA57LDDcthhh9XYp/pM5/vtt98WgaVz584ZOHBgpk2blilTpmTBggXZb7/9tljP5MmTS5MWDhkyZKdrBwCA5q5ZXoKVJBdddFGSDbfhHTp0aFasWLHJ8nnz5mX06NFJko4dO+acc84pvEYAAGhqyu4MSH054YQTcvrpp+fuu+/Oo48+mn79+uW8885LZWVlnnrqqYwfP74USq6++up069atkSsGAIDy12wDSJLcdtttWbFiRaZPn54XX3wxl1xyySbLKyoqMmbMmNLvSQAAgJ3TbC/BSpL27dtn2rRpuf3223P88cene/fuadWqVXr27JnBgwdn1qxZGTNmTGOXCQAATUaTPAOy//77p6qqaof6VlRUZNiwYRk2bFgDVwUAADTrMyAAAECxBBAAAKAwAggAAFAYAQQAACiMAAIAABRGAAEAAAojgAAAAIURQAAAgMIIIAAAQGEEEAAAoDACCAAAUBgBBAAAKIwAAgAAFEYAAQAACiOAAAAAhRFAAACAwgggAABAYQQQAACgMAIIAABQGAEEAAAojAACAAAURgABAAAKI4AAAACFEUAAAIDCCCAAAEBhBBAAAKAwAggAAFAYAQQAACiMAAIAABRGAAEAAAojgAAAAIURQAAAgMIIIAAAQGEEEAAAoDACCAAAUBgBBAAAKIwAAgAAFEYAAQAACiOAAAAAhRFAAACAwgggAABAYQQQAACgMAIIAABQGAEEAAAojAACAAAURgABAAAKI4AAAACFEUAAAIDCCCAAAEBhBBAAAKAwAggAAFAYAQQAACiMAAIAABRGAAEAAAojgAAAAIURQAAAgMIIIAAAQGEEEAAAoDACCAAAUBgBBAAAKIwAAgAAFEYAAQAACiOAAAAAhRFAAACAwgggAABAYQQQAACgMAIIAABQmLIPIK+//nouv/zyfOhDH0rHjh3ToUOH9OnTJxdeeGHmzZu33fFVVVWZNGlSjjvuuHTt2jVt27ZN7969M3z48MyePbuAPQAAgOajZWMXsDNmzJiRz3/+81myZMkmzz/33HN57rnncv311+eKK67IP/zDP2x1/KpVqzJ48OBMnTp1k+fnz5+f+fPn54477shVV12VSy+9tMH2AQAAmpOyDSC///3vc8opp2T16tVJkk9/+tP55Cc/mU6dOuV3v/tdbrzxxqxatSrf+ta30qlTp1xwwQVbrGPEiBGl8NGnT5+MHDkyPXr0yJNPPpkbbrghb7/9di677LLss88+OfPMMwvdPwAAaIoqqqqqqhq7iLo49thjM2vWrCTJ+PHjc+65526y/I9//GM+8pGPZNmyZenUqVNefvnldOrUqbR8xowZOeGEE5IkAwYMyNSpU9O2bdvS8meffTbHHHNMFi9enK5du+bFF19Mx44dG2x/fvOb32TYsGGlx3fccUcOP/zwBtse0LBeeeWVJMm+++7byJUAUF98X6sfZfkbkHnz5pXCxwknnLBF+EiS97///bnkkkuSJMuXL8999923yfKxY8cmSVq2bJmJEyduEj6SDWdExo0blyRZvHhxJkyYUO/7AQAAzU1ZBpAlS5ZkwIAB6dGjR4YMGbLNfv369Su1X3rppU3GP/DAA0mSQYMGpXfv3lsdf8YZZ2TPPfdMktx11131UToAADRrZfkbkI997GN58MEHt9tv4yUQSbLXXnuV2rNmzcr69euTJAMHDtzm+BYtWqR///6566678vjjj2fp0qXZY489dqJyAABo3sryDMiOWLJkSa655pokSfv27XPiiSeWls2ZM6fU7tu3b43rOfjgg5NsuF3vM8880wCVAgBA89GkAsjq1avz/PPP51/+5V9y6KGH5g9/+EOS5F//9V/TrVu3Ur/58+eX2vvvv3+N66ysrCy1q1/GBQAA1F5ZXoK1Nb/97W+3uAvBXnvtlX//93/f4nciixYtKrWrB5Ot6dKlS6m9ePHieqgUAACaryZzBuTll1/e4rnFixfnzjvvzB//+MdNnl+5cmWpvfndrzZXfXn1cQAAQO01mQDSvXv3jBs3LnfeeWfGjh2bfv365Z133sk999yTI444Ik8++WSp77vvvltqt2nTpsb1Vl9efRwAAFB7TeYSrKOPPjpHH3106fFFF12U8847LzfddFPeeuutDBs2LHPmzMluu+2Wdu3alfqtXbu2xhCyZs2aUrt169YNUzwAADQTTeYMyOZatGiR66+/Ph/4wAeSJM8991ymT5+eJOnQoUOp3+rVq2tcT/Xl1YMLAABQe002gCTJbrvtlq985Sulx4888kiSpHPnzqXnlixZUuM6qi/v3r17/RYIAADNTJMOIEly4IEHltpvvPFGkuSAAw4oPbdgwYIax1f/cXuvXr3quToAAGheyjKA3HrrrRk8eHAOO+ywLFu2rMa+1S+h6tixY5L/m1wwSebOnVvj+I3LKyoqcsghh9SxYgAAICnTAPLMM8/kpz/9aZ566qk8+OCDNfZ9/PHHS+2DDjooSXLUUUelVatWSZKHHnpom2PXrVuXhx9+OEnSr1+/TS7dAgAAaq8sA8igQYNK7euvv36b/ZYsWZKJEycmSVq1apVPf/rTSTb8BmTgwIFJkilTpmzzMqzJkyeXJi3cfDJDAACg9soygAwYMCD9+vVLksycOTNjx47dos/y5cvz2c9+thQgRo4cmb333ru0/KKLLkqy4Ta8Q4cOzYoVKzYZP2/evIwePTrJhku3zjnnnAbZFwAAaE7Kch6QFi1a5Mc//nGOOeaYrFy5Mn/3d3+XGTNm5PTTT8/uu++euXPnZuLEiVm4cGGS5EMf+lD+5V/+ZZN1nHDCCTn99NNz991359FHH02/fv1y3nnnpbKyMk899VTGjx9fCiVXX311unXrVvh+AgBAU1OWASRJDjvssPzyl7/MkCFD8sYbb2T69OmleT6qGzRoUH7yk5+kffv2Wyy77bbbsmLFikyfPj0vvvhiLrnkkk2WV1RUZMyYMRk1alSD7QcAADQnZXkJ1kb9+/fPc889l+9973s54ogj0qlTp7Ru3Tr77rtvhgwZkqlTp+a+++7b5o/H27dvn2nTpuX222/P8ccfn+7du6dVq1bp2bNnBg8enFmzZmXMmDHF7hQAADRhZXsGZKM99tgjl19+eS6//PI6ja+oqMiwYcMybNiweq4MAADYXFmfAQEAAMqLAAIAABRGAAEAAAojgAAAAIURQAAAgMIIIAAAQGEEEAAAoDACCAAAUBgBBAAAKIwAAgAAFEYAAQAACiOAAAAAhRFAAACAwgggAABAYQQQAACgMAIIAABQGAEEAAAojAACAAAURgABAAAKI4AAAACFEUAAAIDCCCAAAEBhBBAAAKAwAggAAFAYAQQAACiMAAIAABRGAAEAAAojgAAAAIURQAAAgMIIIAAAQGEEEAAAoDACCAAAUBgBBAAAKIwAAgAAFEYAAQAACiOAAAAAhRFAAACAwgggAABAYQQQAACgMAIIAABQGAEEAAAojAACAAAURgABAAAKI4AAAACFEUAAAIDCCCAAAEBhBBAAAKAwAggAAFAYAQQAACiMAAIAABRGAAEAAAojgAAAAIURQAAAgMIIIAAAQGEEEAAAoDACCAAAUBgBBAAAKIwAAgAAFEYAAQAACiOAAAAAhRFAAACAwgggAABAYQQQmpV169Zl3bp1jV0GAECzJYDQrCxcuDALFy5s7DIAAJotAQQAACiMAAIAABRGAAEAAAojgAAAAIVpEgFk6dKl+cEPfpCjjz46Xbt2TevWrbPnnnvmuOOOy4033pi1a9fWOL6qqiqTJk3Kcccdl65du6Zt27bp3bt3hg8fntmzZxe0FwAA0PS1bOwCdtbMmTPz+c9/PosWLdrk+UWLFuXBBx/Mgw8+mHHjxuXnP/95evfuvcX4VatWZfDgwZk6deomz8+fPz/z58/PHXfckauuuiqXXnppg+4HAAA0B2UdQH7/+9/n5JNPzsqVK5MkJ5xwQk499dR07do1L730Um699dbMnTs3c+fOzQknnJDZs2enc+fOm6xjxIgRpfDRp0+fjBw5Mj169MiTTz6ZG264IW+//XYuu+yy7LPPPjnzzDOL3kUAAGhSyjqAfP3rXy+Fj/Hjx+fcc8/dZPlFF12UL33pS7njjjvypz/9Kd/73vdyzTXXlJbPmDEjkyZNSpIMGDAgU6dOTdu2bZMkQ4cOzZe//OUcc8wxWbx4cS688MJ85jOfSceOHQvaOwAAaHrK9jcgzz33XH71q18lSU499dQtwkeStGzZMhMnTszee++dJLn55ps3mQV77Nixm/TbGD426tOnT8aNG5ckWbx4cSZMmNAg+wIAAM1F2QaQBx54oNQ+66yzttmvbdu2Oemkk5Js+LH6H//4xyTJkiVLSusYNGjQVn8fkiRnnHFG9txzzyTJXXfdVS+1AwBAc1W2AWS33XbLIYcckk6dOuWAAw6osW+XLl1K7aVLlyZJZs2alfXr1ydJBg4cuM2xLVq0SP/+/ZMkjz/+eGk8AABQe2UbQM4777zMmTMnb731Vvr27Vtj33nz5pXaXbt2TZLMmTOn9Nz2xh988MFJNtyu95lnnqlryQAA0OyVbQDZUa+99lqmTZuWJOnWrVve9773Jdlwm92N9t9//xrXUVlZWWq/9NJL9V4jAAA0F00+gHzjG98oTUT4+c9/Pi1abNjl6vOGdOvWrcZ1VL+Ea/HixQ1QJQAANA9NOoDceOONmTx5cpKkQ4cO+fu///vSso23702yxd2vNld9efVxAABA7TTZAPKzn/0sX/3qV0uPx48fX7odb5K8++67pXabNm1qXFf15dXHAQAAtdMkA8g999yTIUOGlOb8GD169BazmLdr167U3niJ1rasWbOm1G7dunU9VgoAAM1LkwsgP/rRj3LGGWfknXfeSZIMHz4811577Rb9OnToUGqvXr26xnVWX149uAAAALXTpALId7/73YwYMaJ0mdSXv/zl/PjHP05FRcUWfTt37lxqL1mypMb1Vl/evXv3+ikWAACaoSYRQNavX59Ro0bliiuuKD03evToTJw4sXTXq81Vn7xwwYIFNa7/5ZdfLrV79eq1c8UCAEAzVvYBZP369fniF7+Ym266qfTc9773vfzbv/3bVs98bLRxcsEkmTt3bo3b2Li8oqIihxxyyE5WDAAAzVfZB5Dzzz8/d9xxR5KkRYsWGT9+fC6//PLtjjvqqKPSqlWrJMlDDz20zX7r1q3Lww8/nCTp16/fJpduAQAAtVPWAeSWW27JjTfemGRD+Ljlllty7rnn7tDYzp07Z+DAgUmSKVOmbPMyrMmTJ5cmLRwyZEg9VA0AAM1X2QaQ119/PaNHjy49Hjt2bM4666xareOiiy5KsuE2vEOHDs2KFSs2WT5v3rzSNjp27JhzzjlnJ6sGAIDmrWVjF1BX//Zv/5bly5cnSfbbb7/07t07P/vZz7Y77rDDDst+++2XJDnhhBNy+umn5+67786jjz6afv365bzzzktlZWWeeuqpjB8/vhRKrr766nTr1q3B9gcAAJqDsg0gt956a6m9YMGCnHbaaTs07uabb86XvvSl0uPbbrstK1asyPTp0/Piiy/mkksu2aR/RUVFxowZk1GjRtVL3QAA0JyV5SVYb775ZhYuXFgv62rfvn2mTZuW22+/Pccff3y6d++eVq1apWfPnhk8eHBmzZqVMWPG1Mu2YKN169Zl3bp1jV0GAEDhyvIMSLdu3VJVVVVv66uoqMiwYcMybNiwelsn1GRjgN53330buRIAgGKV5RkQAACgPAkgAABAYQQQAACgMAIIAABQGAEEAAAojAACAAAURgABAAAKI4AAAACFEUAAAIDCCCAAAEBhBBAAAKAwAggAAFAYAQQAACiMAAIAABRGAAEAAAojgAAAAIURQGAXsm7duqxbt66xywAAaDACCOxCFi5cmIULFzZ2GQAADUYAAQAACiOAAAAAhRFAAACAwgggAABAYQQQAACgMAIIAABQGAEEAAAojAACAEAhTLhLIoAAAFAQE+6SCCAAAECBBBAAAKAwAggAAFAYAQQAACiMAAIAABRGAAEAAAojgAAAAIURQAAAgMIIIAAAQGEEEAAAoDACCAAAUBgBBAAAKIwAAgAAFEYAAQAACiOAAAAAhRFAAACAwgggAABAYQQQAACgMAIIAABQGAEEAAAojAACAAAURgABAAAKI4AAAACFEUAAAIDCCCAAAEBhBBAAAKAwAggAAFAYAQQAACiMAAIAABRGAAEAAAojgAAAAIURQAAAgMIIIAAAQGEEEAAAoDACCAAAUBgBBAqybt26VFVVNXYZAACNSgCBgixcuDDr1q1r7DIAABqVAAIAABRGAAEAAAojgAAAAIURQAAAgMI0yQDy1a9+NRUVFbniiit2qP8vf/nLnHzyyenRo0fatGmTysrKnH766XnggQcatlAAAGhmWjZ2AfVt5syZueGGG3ao7/r16zNq1KhMnDhxk+dfeeWVvPLKK7nnnntywQUX5LrrrktFRUVDlAtb2HPPPRu7BACABtOkAshvf/vbnHbaaVm/fv0O9f/2t79dCh+VlZU5//zzs//+++e5557L9ddfn0WLFuU//uM/0qNHj1x++eUNWTqUtGjRJE9MAgAkaUIB5L777ssXvvCFLF++fIf6P/fcc/nnf/7nJMkhhxySWbNmZY899igtHzVqVI499tj86U9/yve+972cddZZ6dWrV4PUDgAAzUXZ/6/WNWvWZMyYMTnppJOybNmyHR537bXXliaFGz9+/CbhI0l69uyZ2267LUmydu3aXHPNNfVWMwAANFdlHUAeeOCB9OnTJ1deeWXWr1+f97znPbn44ou3O279+vX56U9/miTp27dvjjnmmK32O+qoo/LhD384SXL33Xenqqqq/ooHAIBmqKwDyO23354XX3wxSXL44YfniSeeyMknn7zdcc8880yWLFmSJBk4cGCNfQcMGJAkee211/L000/vZMUAANC8lXUASZLu3bvnhz/8YR577LEcfPDBOzRmzpw5pXbfvn1r7Ft9nb///e/rViQAAJCkzH+EfsEFF2T8+PFp165drcbNnz+/1N5///1r7FtZWVlqv/TSS7XaDgAAsKmyDiCHH354ncYtWrSo1O7WrVuNfbt06VJqL168uE7bAwAANij7S7DqYuXKlaV227Zta+xbfXn1cQAAQO01ywDy7rvvltpt2rSpsW/15dXHAQAAtdcsA0j134ysXbu2xr5r1qwptVu3bt1gNVF369atK83p0pDb2Hgb5iK2B0DNfBY3Xd7bpq9ZBpAOHTqU2qtXr66xb/Xltf2xO8VYuHBhFi5c2ODb2PhhWMT2AKiZz+Kmy3vb9DXLANK5c+dSe+N8INtSfXn37t0bqiQAAGgWmmUAOeCAA0rtBQsW1Nj35ZdfLrV79erVYDUBAEBz0CwDSPXJBefOnVtj3+rLDz300AarCQAAmoNmGUAOPPDA9OjRI0ny0EMP1dh35syZSTbMB7K9WdMBAICaNcsA0qJFi5x22mlJkt/85jd57LHHttrvkUceyZNPPpkk+dznPpcWLZrlywUAAPWm2X6j/trXvpaWLTdMBP/FL34xf/nLXzZZ/tprr2X48OFJkt122y0XXnhh0SUCAECT02wDyMEHH5zRo0cnSf74xz+mX79+ufLKKzN58uR897vfzYc+9KG88MILSZK/+7u/S58+fRqzXABgM3vuuWdjl0AdeN9o2dgFNKZ//ud/zptvvplbb701b7zxRsaMGbNFnxEjRuSqq65qhOoAgJq4NLo8ed9o1kfAbrvtlltuuSVTp07NKaeckr322iutWrVKt27d8ulPfzq/+MUvMmHCBH8oBTHz6dZ5XQBq1lw/Jxtzv+tj2831faMJngH5+Mc/nqqqqlqNOfHEE3PiiSc2UEXsqI2znu67776NXMmuxesCULPm+jnZmPtdH9turu8bzfwMCAAAUCwBBAAAKIwAAgAAFEYAAQAACiOAAAAAhRFAaBLKcVKjcqwZAGBnCSA0CeU4V0s51gwAsLN8A6JJWbduXWkemOrtxqhjRydXasw6qR2TZgHAzhNAaFIWLlxY+oJYvd0YdWycYGlH+vpSWx5q874CAFsngAAAAIURQAAAgMIIIAAAQGEEEAAAoDACCAAAUBgBBOrAJIIAAHUjgEAdmEQQAKBufItim5P3NadJ1+q6r83pNQJoLCZsLW/ePzYngLDNyfua06Rrdd3X5vQaATQWE7aWN+8fmxNAAACAwgggAABAYQQQAACgMAIIAABQGAEEAAAojADCLqUxJ/ir722brBAAYEsCCLuUxpzgr763bbJCAIAt+YZE4XZk8r7t9WnISY2a6+SCzXW/ARqDz1yaMwGEwu3I5H3b69OQkxo118kFm+t+AzQGn7k0ZwIIAABQGAEEAAAojAACAAAURgABAAAKI4AAAACFEUBoFE1pwkEAoG78m9w8CSA0iqY04SAAUDf+TW6evOvUSvUJAOtjEqWGnFBwV952fSt6QqvmNIFWUzpOaF7HLlt/v2t7DGzvM6Ahj6m6rrs2n1v1XX8RNVP+BBBqpfoEgPUxiVJDTii4K2+7vhU9oVVzmkCrKR0nNK9jl62/37U9Brb3GdCQx1Rd112bz636rr+Imil/AggAAFAYAQQAACiMAAIAABRGAAEAAAojgAAAAIURQNgpTXUCodrsVxGvwY5so6m+FwA7w2djw/HaUlcCCDulqU4gVJv9KuI12JFtNNX3AmBn+GxsOF5b6sqRQ73Y1sRDzWnSryImUdqR17M+69h84snmMklUc9pX2FX5Oywf3itqSwChXmxr4qHmNOlXEZMo7cjrWZ91bD7xZHMJk81pX2FX5e+wfHivqC0BBAAAKIwAAgAAFEYAAQAACiOAAAAAhRFAAACAwgggAABAYQQQ6k19z4hqhtWt87qUD+/Vrsn7Ars+f6dNmwBCvanvGVHNsLp1Xpfy4b3aNXlfYNfn77Rp8+5S78yIWoy6vs6NOTt9Xbe9rXHlMlP75nU29OvfmO9xXZVjzU1J0a+/9xuaNwGEemdG1GLU9XVuzNnp67rtbY0rl5naN6+zoV//xnyP66oca25Kin79vd/QvAkgAABAYQQQAACgMAIIAABQGAEEAAAojAACAAAURgChwZlMCDblb2LrvC6Nq+jXvym933Xdl6b0GlTXVPeL+iOA0OBMJgSb8jexdV6XxlX069+U3u+67ktTeg2qa6r7Rf1xhFA2duWJ5mjeipgUsqgJDLc2sWNN7Yaoqb63sSt/dtR2/7a2L/X9PjTE+mp6/XeVSQnrY2LT7a1jW+ut7WdBEZPQ7sp/N5Q/AYSysStPNEfzVsSkkEVNYLi1iR1rajdETfW9jV35s6O2+7e1fanv96Eh1lfT67+rTEpYHxObbm8d21pvbT8LipiEdlf+u6H8CSAAAEBhBBAAAKAwAggAAFAYAQQAACiMAJKkqqoqkyZNynHHHZeuXbumbdu26d27d4YPH57Zs2c3dnkAANBktGzsAhrbqlWrMnjw4EydOnWT5+fPn5/58+fnjjvuyFVXXZVLL720kSpsXCYTapq29b7W5v1urhNvlXv99c3r0TSV46SERXx+NYb6rLU+PvuhPjT7ADJixIhS+OjTp09GjhyZHj165Mknn8wNN9yQt99+O5dddln22WefnHnmmY1cbfFMJtQ0bet9rc373Vwn3ir3+uub16NpKsdJCYv4/GoM9VlrfXz2Q31o1kfcjBkzMmnSpCTJgAED8uSTT+bCCy/M0KFDc/XVV+eJJ55I165dkyQXXnhhVqxY0ZjlQp3VNPnVzk40tb3JrXZk27WZsKu2tW1vQrv6mHCsHO0qk7/VVX3WXx+TLu6KkwHW9W+zrnXU90R+dVWbdTTnv/+G+jehCEVMiErDatYBZOzYsUmSli1bZuLEiWnbtu0my/v06ZNx48YlSRYvXpwJEyYUXiPUh5omv9rZD+7tTW61I9uuzYRdta1texPa1ceEY+VoV5n8ra7qs/76mHRxV5wMsK5/m3Wto74n8qur2qyjOf/9N9S/CUUoYkJUGlazDSBLlizJAw88kCQZNGhQevfuvdV+Z5xxRunayLvuuquw+gAAoClqtgFk1qxZWb9+fZJk4MCB2+zXokWL9O/fP0ny+OOPZ+nSpYXUBwAATVGzDSBz5swptfv27Vtj34MPPjjJhtv1PvPMMw1aFwAANGXNNoDMnz+/1N5///1r7FtZWVlqv/TSSw1UEQAANH0VVeVwu4MGcOqpp+bee+9NkixdujSdO3feZt///u//zmc/+9kkybXXXpsLL7yw3uv51a9+lXPOOaf0+Dvf+U4OPPDAet/O1ixatChdunTJbrvtts12bfrW9zpse9ffdpJ07969zsfXjmx7a9uo73U0xmte02u3eZ312Xdrfepjv+uqtq9RbV+PIo7zHa1nR8bV9nitzTq2tq76Okab0uda0duuzfvdWJ9rjfl+17Ttrb0GDeX555/PlVdeWXo8YcKEHHvssYVsuylptvOArFy5stTe/O5Xm6u+vPq4+vTyyy9v8rj6wQ0AwK5n8+9v7JhmewnWu+++W2q3adOmxr7Vl1cfBwAA1E6zDSDt2rUrtdeuXVtj3zVr1pTarVu3brCaAACgqWu2l2B16NCh1F69enWNZ0FWr15dalcPLvVpwIABmzzeb7/9GmxbAADU3qpVq7JgwYLS482/v7Fjmm0Aqf6j8yVLlmT33XffZt8lS5aU2g31I6eePXtm2LBhDbJuAADYVTTbS7AOOOCAUrt6kt2a6j8w6tWrV4PVBAAATV2zDSAbJxdMkrlz59bYd+PyioqKHHLIIQ1aFwAANGXNNoAcddRRadWqVZLkoYce2ma/devW5eGHH06S9OvXr8b5QgAAgJo12wDSuXPnDBw4MEkyZcqUbV6GNXny5NIkN0OGDCmsPgAAaIqa7UzoSTJ9+vR88pOfTJL87d/+baZNm5aOHTuWls+bNy/HHntsFi9enI4dO+bPf/5zunXr1ljlAgBA2WvWASRJPve5z+Xuu+9OkvTu3TvnnXdeKisr89RTT2X8+PFZsWJFkuSGG27IqFGjGrNUAAAoe80+gKxcuTKnnXZapk+fvtXlFRUVGTNmTMaMGVNwZQAA0PQ0+wCSJFVVVZk0aVJuvfXW/O53v8uyZcvSrVu3fOxjH8vo0aNz9NFHN3aJAADQJAggAABAYZrtXbAAAIDiCSAAAEBhBBAAAKAwAggAAFAYAQQAACiMAAIAABRGAAEAAAojgAAAAIURQAAAgMIIIM1IVVVVJk2alOOOOy5du3ZN27Zt07t37wwfPjyzZ89u7PIoQw19TPXu3TsVFRU79N9f/vKXetgjmquvfvWrqaioyBVXXNHYpdAE1Ofx9IlPfGKHPwcfe+yxnS8eCtCysQugGKtWrcrgwYMzderUTZ6fP39+5s+fnzvuuCNXXXVVLr300kaqkHLT0MfU8uXL89JLL9VHqVCjmTNn5oYbbmjsMmgi6vt4euaZZ+ptXbCrEECaiREjRpS+KPbp0ycjR45Mjx498uSTT+aGG27I22+/ncsuuyz77LNPzjzzzEaulnLQ0MfUM888k6qqqiTJt771rRx++OE19t9jjz1qvxM0e7/97W9z2mmnZf369Y1dCk1AfR9Pr776ahYvXpwkOeecc3LiiSfW2P/AAw+sl+1Cg6uiyZs+fXpVkqokVQMGDKhatWrVJsvnzZtX1bVr16okVV27dq1avnx5I1VKuSjimLr++utL23j++efrq3Qo+eUvf1nVuXPn0nGWpGrMmDGNXRZlqiGOp1/+8peldd1///31UyjsAvwGpBkYO3ZskqRly5aZOHFi2rZtu8nyPn36ZNy4cUmSxYsXZ8KECYXXSHkp4ph6+umnkyRt27bN+973vp2sGP7PmjVrMmbMmJx00klZtmxZY5dDmWvI42nj52CS9O3bt17XDY1JAGnilixZkgceeCBJMmjQoPTu3Xur/c4444zsueeeSZK77rqrsPooP0UdUxv/4T344IPTooWPKurHAw88kD59+uTKK6/M+vXr8573vCcXX3xxY5dFmWro42nj52CXLl2y995719t6obH5V72JmzVrVula1IEDB26zX4sWLdK/f/8kyeOPP56lS5cWUh/lp4hjqqqqKnPmzEmSHHrooTtRLWzq9ttvz4svvpgkOfzww/PEE0/k5JNPbuSqKFcNfTxtDCA+B2lqBJAmbuOXuGT7p28PPvjgJBu+/LnrBttSxDH10ksvZfny5ZtsY/ny5Zk1a1buu+++/O53v8u6detqWzokSbp3754f/vCHeeyxx0rHKNRVQx1P77zzTp5//vkk//c5uHLlyvz617/Offfdl9mzZ2ft2rX1tj0okrtgNXHz588vtffff/8a+1ZWVpbabn/KthRxTFW/7rlFixb57Gc/mylTpmwSOrp27ZoLLrggl156adq1a7fD66Z5u+CCCzJ+/HjHDPWiIY+nZ599Nu+8806SZPfdd8/ZZ5+dyZMnZ/Xq1aU+HTt2zNlnn50rrrjCnQApKwJIE7do0aJSu1u3bjX27dKlS6m98bZ/sLkijqnqAeQb3/jGVvssXrw43/3ud3Pfffdl6tSp260Fkmz3ds5QGw15PFX/HPzHf/zHrfZZsWJFrrvuuvzyl7/M/fffn7/5m79psHqgPrkEq4lbuXJlqb35nYo2V3159XFQXRHHVPV/eFu3bp3LLrsszz77bFavXp3XXnstEyZMSM+ePZMkTzzxRM444wzzOABNSvXPwYqKiowaNSpPPfVUVq5cmUWLFmXy5Mk54IADkiR/+tOfcvLJJ/u3m7IhgDRx7777bqndpk2bGvtWX159HFRXxDH16quvJtkQYB588MH84Ac/yEEHHZQ2bdqkZ8+eGTFiRGbPnp199903yYaZh++8887a7AbALu2VV15JsiF83HnnnbnhhhvywQ9+MO3atUu3bt1yxhln5IknnsgHP/jBJMm8efPy7//+741YMew4AaSJq35d6vZ+rLZmzZpSu3Xr1g1WE+WtiGPqkUceyfLly/OHP/whH/vYx7baZ5999sl1111Xemz+GqApmTRpUv7617/mhRdeyODBg7faZ/fdd8+Pf/zj0mOfg5QLAaSJ69ChQ6ld/YdrW1N9uR9osi1FHVMdO3bc5EfsW3PKKaekU6dOSZJHH33UZVhAk9K+ffttzrW00Yc+9KEccsghSZIXX3yxdAYZdmUCSBPXuXPnUnvJkiU19q2+vHv37g1VEmVuVzqmdtttt7z//e9PsuFsi5snAM3RQQcdVGq/9tprjVgJ7BgBpInb+AO1JFmwYEGNfV9++eVSu1evXg1WE+VtVzumWrVq1SDrBSgXPgcpN27D28RVnxRp7ty5pZmpt2bu3LlJNvzgbePpXNhcQx9TL774YqZMmZLXX389Rx99dD796U/X2H/j5QatW7fe5La/AOXqzTffzOTJk/P666/ngAMOyFlnnVVj/+qXXW28QyDsypwBaeKOOuqo0v8Zeeihh7bZb926dXn44YeTJP369dvkMhuorqGPqddeey0XXnhhfvCDH+Smm26qse8f/vCH0lmWI444IrvtttsObQNgV7ZmzZp87Wtfy/e///2MHTu2xr5vvfVWfvOb3yTZcHOOjXcHhF2ZANLEde7cOQMHDkySTJkyZZuXzEyePLk0wdyQIUMKq4/y09DH1Ic//OHSD8unTZtW42Ve3//+90vt7f0fQoBysc8++5Qud3366afz2GOPbbPvNddck1WrViXxOUj5EECagYsuuijJhlumDh06NCtWrNhk+bx58zJ69OgkG+48dM455xReI+WlIY+ptm3blvqvXbs2X/jCF7J8+fIt+o0dOzb/+Z//mSR5//vfn+HDh9dpXwB2RV/72tdK7eHDh2/1x+WTJ0/OD37wgyRJ165dS5/NsKvzG5Bm4IQTTsjpp5+eu+++O48++mj69euX8847L5WVlXnqqacyfvz40hfIq6++Ot26dWvkitnV7ewx9T//8z/5xCc+kSTp379//ud//meT5d/+9rfzi1/8Is8//3weeeSRHHLIIRk1alQOOOCALFq0KHfeeWdmzZqVZEPAmTRp0nZnZQfYlcyfP790i91evXpl/vz5myw/99xz81//9V+ZNWtW/vCHP+QDH/hARo4cmQ9+8IN56623MmXKlPziF79IkrRs2TK33HJL9txzz6J3A+pEAGkmbrvttqxYsSLTp0/Piy++mEsuuWST5RUVFRkzZkxGjRrVSBVSbhrymNp9990zY8aMfPazn81vfvObvPLKK/n2t7+9Rb/Kysr85Cc/yeGHH17n/QDYFbVs2TJTpkzJ5z//+dx///1ZvHhx6WxHdV26dMmPfvSjnHTSSY1QJdSNS7Caifbt22fatGm5/fbbc/zxx6d79+5p1apVevbsmcGDB2fWrFkZM2ZMY5dJGWnoY6qysjK//vWvc9ttt+VTn/pU9txzz7Rq1SrdunXL3/7t32bs2LGZN29ejj766HrcK4BdR+fOnXPfffflZz/7WU477bTsvffeadWqVfbYY498+MMfzne/+908++yzOfXUUxu7VKiViqqqqqrGLgIAAGgenAEBAAAKI4AAAACFEUAAAIDCCCAAAEBhBBAAAKAwAggAAFAYAQQAACiMAAIAABRGAAEAAAojgAAAAIURQAAAgMIIIAAAQGFaNnYBAADQ2A488MDt9tlnn30yc+bMAqpp2gQQAACavQsuuGCby6ZMmZIFCxbkqKOOKrCipquiqqqqqrGLAACAXdGvfvWrjBw5Mn379s2kSZPSunXrxi6p7AkgAACwFW+99VYGDRqUlStXZsqUKenVq1djl9Qk+BE6AABsxXXXXZclS5bkq1/9qvBRj5wBAQCAzbzwwgs5+eSTs9dee2XatGkuvapHzoAAAMBmJk6cmHXr1mXUqFHCRz1zBgQAAKp5/fXXM3DgwHTt2jUzZswQQOqZMyAAAFDNvffem3feeSdDhgwRPhqAAAIAANXcf//9SZKTTjqpkStpmgQQAAD4/15//fXMmTMnffv2deerBiKAAADA//fb3/42SXLkkUc2ciVNlwACAAD/3zPPPJMk+cAHPtDIlTRdAggAALuk7373uznwwAMzbty4HepfVVWVn//85/nSl76UI488MoceemgGDBiQSy+9NE8//fQOreOll15Kkuy11151rpuatWzsAgAAYHO//vWvM3ny5B3uv3r16owePTr/8z//s8nzr776al599dX8/Oc/z4UXXpiRI0fWuJ4lS5YkSTp27FjrmtkxAggAALuUOXPm5IILLsj69et3eMy3vvWtUvh473vfmyFDhqRbt26ZO3duJk+enJUrV+aaa65Jjx498pnPfGab66lN6KFuBBAAAHYZDz/8cL75zW/m7bff3uExjzzySH7xi18kST760Y/mpptuSps2bZJsuJXu5z73uXzhC1/IsmXL8o//+I8ZOHBgOnTo0CD1s31+AwIAwE5Zu3btdvu8++67213Hddddl3PPPTfLly+v1fZ//OMfJ0latmyZ73//+6XwsdF73/vefPvb306SLFu2LHfddVet1k/9EkAAAKizO++8M0OHDq3xjMX8+fPz6U9/Or///e+3uvzRRx/Npz71qfzwhz/M+vXr0759+5x99tk7tP1ly5bl0UcfTZIcc8wxqays3Gq/E088MV27dk2STJs2bYfWTcMQQAAAqJOZM2fmiiuuyJw5czJy5MisWrVqiz6vvfZazj777MyfPz/nnHNOXnnllS36TJkypfR83759c9ddd+UTn/jEDtXwm9/8pvRbkY9+9KPb7NeiRYscccQRSZLf//73eeutt3Zo/dQ/AQQAgDr58Ic/nAMOOCDJhgn8zj///E0ux1q0aFG+9KUv5bXXXkuS9O/fP/vss89W19WlS5d85zvfyX/913/lfe973w7X8Mc//rHU3ljLtmxcb1VVVf7whz/s8DaoXwIIAAB1svvuu+fHP/5xevfunWTDpVSjR4/Ou+++m6VLl+bss88uzatx/PHH5wc/+EEqKiq2WM+wYcPy0EMPZdiwYdltt91qVcOrr75aam8r3GxUfW6P6uMolgACAECdde3aNbfeemv23XffJBsuy/rGN76Rr3zlK6WzEx/72Mfyr//6r2nZcus3YD300EPTtm3bOm1/47wdSbLHHnvU2Ldz586l9rJly+q0PXaeAAIAwE7p0aNHbrnllvTo0SPJhh95z507N0ly+OGH5z/+4z/SunXrBtn26tWrS+3N7361ueo1VB9HsQQQAAB2WmVlZW666aa0aPF/Xy+7deuWG2+8Me3atWuw7Va/ve/2Qk715du7LTANRwABAGCnrVu3rnQb3Y3efPPN/OhHP2rQ7Va/dOudd96psW/1H8g31BkZtk8AAQBgp6xfvz6XXXZZpk+fnmTD2ZD27dsnSa6//vpMmDChwba9cTtJsmbNmhr7Vg8g27tci4YjgAAAsFOuuOKKTJkyJUmy995757bbbsv48eNLX/LHjh2bO+64o0G23alTp1J7ez8sr768S5cuDVIP2yeAAABQZ//0T/+UO++8M0nSvXv33HLLLdl7773z0Y9+NOPGjUurVq2SJN/73vdyzz331Pv2999//1J74cKFNfb9y1/+Umrvvffe9V4LO0YAAQCgTn7xi1/k5ptvTrLhFrc333xzevXqVVrev3//jB07Nrvttluqqqpy+eWX54UXXqjXGt773veW2tUnJdyajcsrKiry/ve/v17rYMf9PwdV2nIWnlOBAAAAAElFTkSuQmCC",
      "text/plain": [
       "<Figure size 400x400 with 1 Axes>"
      ]
     },
     "metadata": {
      "image/png": {
       "height": 400,
       "width": 400
      }
     },
     "output_type": "display_data"
    },
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAyAAAAMgCAYAAADbcAZoAAAAOXRFWHRTb2Z0d2FyZQBNYXRwbG90bGliIHZlcnNpb24zLjUuMSwgaHR0cHM6Ly9tYXRwbG90bGliLm9yZy/YYfK9AAAACXBIWXMAAB7CAAAewgFu0HU+AABET0lEQVR4nO3dd5hcZd0//vcmIYUaSKEZIFJDB/OICtIFRaKA0rtiIlJtD6hgQhMLPiiIoUOoQZEHQglSgsiPIkXpIuVJQGCFEAgBU0g2+/sjZr4bSNkku/dkZ16v68p13TPnPvf5nJmzM/POaQ3Nzc3NAQAAKKBTtQsAAADqhwACAAAUI4AAAADFCCAAAEAxAggAAFCMAAIAABQjgAAAAMUIIAAAQDECCAAAUIwAAgAAFCOAAAAAxQggAABAMQIIAABQjAACAAAUI4AAAADFCCAAAEAxAggAAFCMAAIAABQjgAAAAMUIIAAAQDECCAAAUIwAAgAAFCOAAAAAxQggAABAMQIIAABQjAACAAAUI4AAAADFCCAAAEAxAggAAFCMAAIAABQjgAAAAMUIIAAAQDECCAAAUIwAAgAAFCOAAAAAxQggAABAMQIIAABQjAACAAAUI4AAAADFCCAAAEAxAggAAFCMAAIAABQjgAAAAMUIIAAAQDECCAAAUIwAAgAAFCOAAAAAxQggAABAMQIIAABQjAACAAAUI4AAAADFCCAAAEAxAggAAFCMAAIAABQjgAAAAMUIIAAAQDECCAAAUIwAAsACPfLIIznxxBOz7bbbZuWVV063bt2yzDLLZI011sjuu++es88+O+PHj5/vGMOGDUtDQ8N8/3Xq1CnLLLNM1l577eyxxx656KKLMm3atEJrCUAJDc3Nzc3VLgKAJdMDDzyQ733ve3nwwQcX2LdHjx454YQT8qMf/ShdunT5yPRhw4bllFNOWega1l133Vx55ZXZaqutFnpeAJY8H/2GAIAkQ4cOzWmnnZbZ/0/VpUuXbLPNNtliiy3Su3fvTJ8+PS+88ELuvPPOvPnmm5kyZUqGDRuW++67LzfeeGOWXXbZeY49cODA7Lvvvh95/oMPPsjEiRPz/PPP584778zkyZPzwgsvZLfddssDDzyQ9ddfv93WF4AyBBAAPuL73/9+zjrrrMrjIUOGZNiwYVlllVU+0nf69Ok555xzcuKJJ2bGjBm5++67c8ABB+Smm25KQ0PDXMffaKON8r3vfW++NfzrX//Kbrvtlr/97W95++238+1vfzu33Xbb4q0YAFXnHBAA5nD99ddXwkdDQ0OuuOKKnH/++XMNH0my1FJL5bvf/W5GjhxZCRw333xzrrjiisWqY5VVVsmVV15ZeTx69OiMGzduscYEoPoEEAAqpk6dmuOOO67y+Ic//GEOPvjgVs37la98JYcddljl8bBhw7K4pxlutNFGWXfddSuPn3766cUaD4DqE0AAqLjyyivz+uuvJ0n69OmTk046aaHmP+mkkyp7QZZaaqm8+OKLi13TiiuuWGm7bgpAx+ccEAAqrr766kp7v/32S/fu3Rdq/o9//OO58847s/HGG2fllVde7HqmT5+el156Kcmsw8E23XTTxR4TgOoSQABIkkyZMiX3339/5fGOO+64SOPstNNObVVSfvOb32TChAlJki996UtZc80122xsAKpDAAEgSfL3v/89M2bMqDzeeuuti9cwc+bMvPvuu3n++edzySWX5OKLL06SrLnmmjn//POL1wNA2xNAAEiS/POf/6y0u3btmj59+rTbskaMGJERI0a0qu+Xv/zlnHfeefO8ChcAHYuT0AFIkrz//vuVdq9evapYyZzGjh2bMWPGVLsMANqIPSAAJEk6d+5caU+bNq1dlzWvO6HPnDkz77//fl555ZX8+c9/ztixY/Pkk0/mkEMOya233pqrr756jjoB6HgEEACSJD179qy033nnnTQ1NbXbj/3W3Al95syZufTSS3PkkUdmxowZue6667L22mvnjDPOaJeaACjDIVgAJEnWWWedSru5uTlvvvlmFatJOnXqlCOOOCJnnnlm5bn/+Z//ydtvv13FqgBYXAIIAElm3cOj5U3//r//7/9bpHH+8pe/5N57780HH3zQJnUdeeSRWWaZZZLMulP7nXfe2SbjAlAdAggASWbtcWh574877rhjkcY57bTTsv3226dnz545/fTTF7uuZZZZZo69M2PHjl3sMQGoHgEEgIoDDzyw0v7f//3f/Pvf/16o+V977bVKcJkyZUrWW2+9Nqmr5bkoPXr0aJMxAagOAQSAikGDBmXttddOkkyYMGGhT/g++eSTM3369CTJ6quvni996UuLXdPUqVPz/PPPVx6vtdZaiz0mANUjgABQ0aVLl/zsZz+rPP7Zz36WK6+8slXzXnTRRbnssssqj4cNG5bu3bsvdk2XXXZZ5R4lSy+9dHbeeefFHhOA6hFAAJjDV77ylRxxxBFJZl0K95BDDsngwYPz+uuvz7X/hAkTcswxx2TIkCGV5/baa6/KGIvj5ptvzve///3K48GDB1dOSAegY2pobm5urnYRACxZmpqacsghh+Saa66pPNelS5dsu+22GThwYFZcccW89957efLJJ3P33XdnypQplX6DBg3Kdddd95FzNYYNG5ZTTjklybxvRJgk06dPz/jx4/OnP/0pf/vb3yrPDxgwIA899FCWX375tlxVAAoTQACYpwsvvDA//vGP88Ybbyyw77LLLpthw4bl+OOPn+sNDFsGkIX1+c9/PpdddllWWWWVRZofgCWHO6EDME+DBw/OQQcdlJtuuim33XZbnnjiibz66qt577330rVr1/Tu3TtbbLFFdtlllxx00EGLvXeioaEhPXr0yAorrJB11103AwcOzD777JOtttqqjdYIgGqzBwQAACjGSegAAEAxAggAAFCMAAIAABQjgAAAAMUIIAAAQDECCAAAUIwAAgAAFCOAAAAAxQggAABAMQIIAABQjAACAAAU06XaBTBLY2NjxowZU3m8xhprpEePHlWsCACAlqZMmZJXXnml8njHHXfMqquuWsWKOiYBZAkxZsyYnHrqqdUuAwCAhXDggQdWu4QOxyFYAABAMQIIAABQjEOwlhD9+vWb4/GPf/zjrL/++lWqBgCAD/vHP/4xxyHzH/79RusIIEuIpZdeeo7H66+/fgYOHFilagAAWJAP/36jdRyCBQAAFCOAAAAAxQggAABAMQIIAABQjAACAAAUI4AAAADFCCAAAEAxAggAAFCMAAIAABQjgAAAAMUIIAAAQDECCAAAUIwAAgAAFCOAAAAAxQggAABAMQIIAABQjAACAAAUI4AAAADFCCAAAEAxAggAAFCMAAIAABQjgAAAAMUIIAAAQDECCAAAUIwAAgAAFCOAAAAAxQggAABAMQIIAABQjAACAAAUI4AAAADFCCAAAEAxAggAAFCMAAIAABQjgAAAAMUIIAAAQDECCAAAUIwAAgAAFCOAAAAAxQggAABAMQIIAABQjAACAAAUI4AAAADFCCAAAEAxAggAAFCMAAIAABQjgAAAAMUIIAAAQDECCAAAUIwAAgAAFCOAAAAAxQggAABAMQIIAABQjAACAAAUU5MB5KijjkpDQ0OGDRu2SPNPnz49m2++eRoaGrL99tu3aW0AAFDPai6AjBkzJueff/5ijXHGGWfkiSeeaKOKAACA2WoqgDz22GPZc889M3PmzEUe4/HHH89PfvKTNqwKAACYrWYCyOjRo7Pzzjtn0qRJizzG9OnTc9hhh2X69OltWBkAADBbhw8g06ZNy9ChQ7P77rtn4sSJizXW7EOvevbs2Sa1AQAAc+rQAeSuu+7KgAEDcuqpp2bmzJlZZpll8p3vfGeRxmp56NUvf/nLtiwTAAD4jw4dQK666qqMHTs2STJw4MA8/PDDGTRo0EKP0/LQq1122SVf+9rX2rpUAAAgHTyAJEmfPn1y3nnn5aGHHsqGG264SGPMPvRqueWWy0UXXdTGFQIAALN1qXYBi+Poo4/O8OHD06NHj0Ueo+WhVz/72c+yxhprtFV5AADAh3ToADJw4MDFmr/loVfbb799vvnNb7ZRZQAAwNx0+EOwFsfsQ6+WXnrpXHzxxWloaKh2SQAAUNPqNoC0PPTqjDPOyNprr13ligAAoPbVZQBpeejVZz7zmRx77LHVLgkAAOpCXQaQ2Ydede/ePZdcckk6darLlwEAAIqru1/eLQ+9Gjp0aDbYYIMqVwQAAPWj7gLIjTfemOnTpydJfvCDH6ShoWGu/2a79957K88ddthhVaoaAABqQ90FEAAAoHo69H1AFsV+++2XzTfffIH99txzzyTJRhttlNNPPz1J3KQQAAAWU90FkA022GChzvvo3bt39thjj/YrCAAA6ohDsAAAgGIEEAAAoBgBBAAAKEYAAQAAiqm5k9C33377NDc3L/Y4bTEGAAAwJ3tAAACAYgQQAACgGAEEAAAoRgABAACKEUAAAIBiBBAAAKAYAQQAAChGAAEAAIoRQAAAgGIEEAAAoBgBBAAAKEYAAQAAihFAAACAYgQQAACgGAEEAAAoRgABAACKEUAAAIBiBBAAAKAYAQQAAChGAAEAAIoRQAAAgGIEEAAAoBgBBAAAKEYAAQAAihFAAACAYgQQAACgGAEEAAAoRgABAACKEUAAAIBiBBAAAKAYAQQAAChGAAEAAIoRQAAAgGIEEAAAoBgBBAAAKEYAAQAAihFAAACAYgQQAACgGAEEAAAoRgABAACKEUAAAIBiBBAAAKAYAQQAAChGAAEAAIoRQAAAgGIEEAAAoBgBBAAAKEYAAQAAihFAAACAYgQQAACgGAEEAAAoRgABAACKEUAAAIBiBBAAAKCYmgwgRx11VBoaGjJs2LAF9p08eXJ+85vfZKeddkrfvn3TtWvX9OrVK9tss01+8Ytf5P3332//ggEAoE50qXYBbW3MmDE5//zzW9X3iSeeyF577ZX/+7//m+P5t99+O/fff3/uv//+nHPOObnpppuy5ZZbtke5AABQV2oqgDz22GPZc889M3PmzAX2fe2117LrrrvmjTfeSJJ86lOfyn777ZfVVlstjY2NGTlyZB588MG8+uqr2XXXXfPoo49mzTXXbO9VAACAmlYzh2CNHj06O++8cyZNmtSq/j/84Q8r4eOEE07Igw8+mOOOOy577713jj322DzwwAM54YQTkiRvvfVWvvOd77Rb7QAAUC86fACZNm1ahg4dmt133z0TJ05s1TzvvvtuRo4cmSTZcsstc+aZZ86135lnnplPfOITSZIbb7wxEyZMaJOaAQCgXnXoAHLXXXdlwIABOfXUUzNz5swss8wyrdpTcd999+WDDz5Ikhx44IFpaGiYa7+GhoZ85StfSZLMnDkzjzzySNsVDwAAdahDB5CrrroqY8eOTZIMHDgwDz/8cAYNGrTA+aZPn55NN900K664YtZbb7359l1ppZUq7XfeeWfxCgYAgDrX4U9C79OnT4YNG5YhQ4akc+fOefPNNxc4z5577pk999yzVeM/++yzlXavXr0WuU4AAKCDB5Cjjz46w4cPT48ePdpl/ClTpuTaa69NknTu3DkDBw5sl+UAAEC96NABpL0Dwamnnprx48cnSb7whS/McTgWAACw8Dr0OSDtafTo0fn5z3+eZNbej1NPPbXKFQEAQMcngMzFQw89lL333rtyQ8Nhw4Zliy22qHJVAADQ8QkgH3L//fdn1113zb///e8ks05Y/+EPf1jlqgAAoDYIIC3cdttt2XXXXSt3U//c5z6Xa6+9Np06eZkAAKAt+GX9H5dcckm+/OUvV/Z8fP7zn8+oUaPSrVu3KlcGAAC1QwBJcvrpp+eII47IjBkzkiRf/epXc9NNN6V79+5VrgwAAGpL3QeQH/zgBzn55JMrj4cMGZLrrrsuXbt2rWJVAABQm+o6gPz0pz/NT3/608rjH/7whzn//POd8wEAAO2kbn9p33PPPXNc3er000/PGWecUcWKAACg9tVlAJk6dWoOP/zwNDc3J0mOPfbY/OhHP6pyVQAAUPu6VLuAahgxYkRefvnlJMlyyy2XrbfeOjfeeOMC59tggw2ywQYbtHN1AABQu+oygFx++eWV9nvvvZd99923VfMNHTo0w4YNa5+iAACgDtTlIVhPPfVUtUsAAIC6VHN7QLbffvvKuR3z8v777xeqBgAAaKku94AAAADVIYAAAADFCCAAAEAxAggAAFCMAAIAABQjgAAAAMUIIAAAQDECCAAAUIwAAgAAFCOAAAAAxQggAABAMQIIAABQjAACAAAUI4AAAADFCCAAAEAxAggAAFCMAAIAABQjgAAAAMUIIAAAQDECCAAAUIwAAgAAFCOAAAAAxQggAABAMQIIAABQjAACAAAUI4AAAADFCCAAAEAxAggAAFCMAMISqampKU1NTa3u29zcvNDzAQBQngDCEqmxsTGNjY2t7js7dCzMfAAAlCeAAAAAxQggAABAMQIIAABQjAACAAAUI4AAAADFCCAAAEAxAggAAFCMAAIAABQjgAAAAMUIIAAAQDECCAAAUIwAAgAAFCOAAAAAxQggAABAMQIIAABQjAACAAAUI4AAAADFCCAAAEAxAggAAFCMAAIAABQjgAAAAMUIIAAAQDECCAAAUIwAAgAAFFOTAeSoo45KQ0NDhg0b1qr+t912WwYNGpSVV1453bp1S79+/fKVr3wld911V/sWCgAAdaZLtQtoa2PGjMn555/fqr4zZ87MkCFDcvHFF8/x/KuvvppXX301N9xwQ44++uicc845aWhoaI9yAQCgrtRUAHnsscey5557ZubMma3qf/LJJ1fCR79+/fKtb30ra621Vp577rn89re/zfjx4/Ob3/wmK6+8ck466aT2LB0AAOpCzQSQ0aNH54ADDsikSZNa1f+5557Lz372syTJRhttlPvuuy8rrrhiZfqQIUOy7bbb5sUXX8xpp52Wgw8+OGuuuWa71A4AAPWiw58DMm3atAwdOjS77757Jk6c2Or5zj777DQ1NSVJhg8fPkf4SJJVV101V1xxRZLkgw8+yC9/+cs2qxkAAOpVhw4gd911VwYMGJBTTz01M2fOzDLLLJPvfOc7C5xv5syZuf7665MkG2+8cT772c/Otd+nP/3pfOITn0iS/OEPf0hzc3PbFQ8AAHWoQweQq666KmPHjk2SDBw4MA8//HAGDRq0wPmeeuqpvP3220mSnXbaab59d9xxxyTJ66+/nieffHIxKwYAgPrWoQNIkvTp0yfnnXdeHnrooWy44Yatmufpp5+utDfeeOP59m055hNPPLFoRQIAAEk6+EnoRx99dIYPH54ePXos1Hzjxo2rtNdaa6359u3Xr1+l/fLLLy/UcgAAgDl16AAycODARZpv/PjxlXbv3r3n23ellVaqtCdMmLBIywMAAGbp8IdgLYrJkydX2t27d59v35bTW84HAAAsvLoMIDNmzKi0u3XrNt++Lae3nI8lV9++fatdAgAA81CXAaTlOSMffPDBfPtOmzat0u7atWu71UTb6dSpLjdrAIAOoS5/qS277LKV9tSpU+fbt+X0hT3ZvaNramqq3KyxI5pb/W29Tos6XlNTU+W+Mh39dQagY/G9Q7XVZQDp2bNnpT37fiDz0nJ6nz592qukJVJjY2MaGxurXcYim1v9bb1OizpeY2Nj5cO/o7/OAHQsvneotroMIOutt16l/corr8y37z//+c9Ke80112y3mgAAoB7UZQBpeXPBZ555Zr59W07fZJNN2q0mAACoB3UZQNZff/2svPLKSZJ77rlnvn3HjBmTZNb9QBZ013QAAGD+6jKAdOrUKXvuuWeS5NFHH81DDz001373339//vrXvyZJvvrVr7q6EgAALKa6/UV9zDHHpEuXWTeCP+SQQ/Kvf/1rjumvv/56Dj300CRJ586dc/zxx5cuEQAAak7dBpANN9wwxx13XJLkhRdeyGabbZZTTz01I0eOzCmnnJItttgiL730UpLk+9//fgYMGFDNcgEAoCZ0qXYB1fSzn/0sb731VkaMGJE333wzQ4cO/UifI444ImeccUYVqgMAgNpTt3tAklmHVl1++eW59dZb86UvfSmrrLJKllpqqfTu3Ttf/OIXc8stt+Siiy5y7gcAALSRmtsDsv3221fuMN1au+22W3bbbbd2qggAAJjNf+0DAADFCCAAAEAxAggAAFCMAAIAABQjgAAAAMUIIAAAQDECCAAAUIwAAgAAFCOAAAAAxQggAABAMQIIAABQjAACAAAUI4AAAADFCCAAAEAxAggAAFCMAAIAABQjgAAAAMUIIAAAQDECCAAAUIwAAgAAFCOAAAAAxQggAABAMQIIAABQjAACAAAUI4AAAADFCCAAAEAxAggsAfr27VvtEgAAihBAYAnQqZM/RQCgPvjVA4U0NTWlubl5sfsAMG9NTU1pamqqdhnF1et60zEJIFBIY2PjAr8cWtMHgHlrbGxMY2Njtcsorl7Xm45JAAEAAIoRQAAAgGIEEAAAoBgBBAAAKEYAAQAAihFAAACAYgQQAACgGAEEAAAoRgABAACKEUAAAIBiBBAAAKAYAQQAAChGAAEAAIoRQAAAgGIEEAAAoBgBBAAAKEYAAQAAihFAAACAYgQQAACgGAEEAAAoRgABAACKEUAAAIBiBBAAAKAYAQQAAChGAAEAAIoRQJJMmzYtw4cPz/bbb5+VVlopXbt2zSqrrJJBgwbl+uuvT3Nzc7VLBACAmtCl2gVU22uvvZbdd989jz/++BzPv/HGG7nllltyyy23ZLfddsvvfve7LLPMMtUpEgAAakRd7wGZPn36HOFjnXXWyU9+8pNcddVVOfnkk9OnT58kyW233ZYDDjigipUCAEBtqOs9IJdffnklfAwaNCjXXXddevToUZl+7LHHZuedd84TTzyRUaNG5Y477sguu+xSpWoBAKDjq+s9IDfccEOSpFOnTrngggvmCB9J0rt37/z617+uPP7DH/5QtD4AAKg1dR1Axo4dmyTp06dPVl111bn22WqrrSrtl19+uUhdAABQq+o6gKywwgpJkjfffDPvv//+XPu0DB2zzwkBAAAWTV0HkE9+8pNJkubm5px11llz7fPzn/+80t51112L1AUAALWqrgPI8ccfn2WXXTZJctppp+X444/Pc889l8mTJ+epp57KIYcckksvvTRJsv3222f//fevZrkAANDh1XUAWXvttTN69OisvvrqmTlzZn79619nwIABWWaZZbLpppvmyiuvTNeuXXPMMcfk1ltvTefOnatdMgAAdGh1HUCSZJtttsm1116bXr16zXV6z549s/7666dLl7q+YjEAALSJuv5VPX369BxyyCEZOXJkkmTHHXfMXnvtlT59+mTs2LG54oor8uyzz+boo4/O9ddfn9tuu+0jl+oFAABar64DyAEHHJDrr78+SXLuuefm6KOPnmP69773vXzrW9/KhRdemD/96U/52te+lmuvvbYapbarpqamJHGIGQAdRlNTUzp16pSGhoY52sCSr24Pwbr33nsr4ePQQw/9SPhIZv0g/+1vf1u5WtbIkSPzzDPPFK2zhMbGxjQ2Nla7DABotcbGxsp/oLVsA0u+ug0gs++CniRHHXXUPPt17tw5xxxzTOXxrbfe2q51AQBALavbAPLCCy9U2ptsssl8+26xxRaV9uy7pwMAAAuvbgNIc3NzpT116tT59u3U6f+9TM6TAACARVe3AeRjH/tYpf3oo4/Ot+/TTz9daa+55prtVhMAANS6ug0gO+20U6V9zjnnzLNfc3NzzjvvvMrjXXbZpV3rAgCAWla3AWSvvfbKWmutlSS5+eabc8YZZ3ykT3Nzc0444YTce++9SWaFj80226xkmQAAUFPq9j4gXbt2zYgRI7Lzzjtn+vTpOemkk3LjjTdm//33z8c+9rG89tprufrqq/PYY48lSfr06ZOLLrqoylUDAEDHVrcBJEm23Xbb3Hrrrdl///0zYcKEPProo3M9H2S99dbLTTfdlDXWWKMKVQIAQO2o20OwZvvc5z6XF198MT/5yU/ymc98JiuuuGK6dOmSPn36ZOedd87w4cPz1FNPZYMNNqh2qQAA0OHV9R6Q2Xr27Jkf/OAH+cEPflDtUgAAoKbV/R4QAACgHAEEAAAoRgABAACKEUAAAIBiBBAAAKAYAQQAAChGAAEAAIoRQAAAgGIEEAAAoBgBBAAAKEYAAQAAihFAAACAYgQQAACgGAEEAAAoRgABAACKEUAAAIBiBBAAAKAYAQQAAChGAAEAAIoRQAAAgGIEEAAAoBgBBAAAKEYAAQAAihFAAACAYgQQAACgGAGEmta3b99qlwAAQAsCCDWtUyebOADAksSvMzqkpqamNDc3t9vYTU1NVVn2otRDOQvzXnjfAGDuBBA6pMbGxnb7cdfY2JjGxsaqLHtR6qGchXkvvG8AMHcCCAAAUIwAAgAAFCOAAAAAxQggAABAMQIIAABQjAACAAAUI4AAAADFCCAAAEAxAggAAFCMAAIAABQjgAAAAMUIIAAAQDECCAAAUIwAAgAAFCOAAAAAxQggAABAMQIIAABQjAACAAAUI4AAAADFCCAAAEAxAggAAFCMAAIAABQjgAAAAMUIIAAAQDECCAAAUIwA8h/PPfdcjj322GywwQZZdtlls+yyy2azzTbLj370o4wfP77a5QEAQE3oUu0ClgRnn312TjzxxHzwwQdzPP/kk0/mySefzMUXX5ybbropn/rUp6pUIQAA1Ia6DyDnnntuvvOd7yRJll566Xz961/PJz/5yUyePDkjR47MPffckzfffDO77bZbnn322ayyyipVrhgAADquuj4Ea9y4cTnhhBOSJH369Mlf/vKXnHPOOTnooIMyePDgjBkzJscee2yS5J133skZZ5xRzXIBAKDDq+sAcsopp2TKlClJkt/97nfZeOONP9Ln5z//efr06VPpAwAALLq6PQRr2rRpueGGG5IkX/rSl7L99tvPtV+3bt0ybNiwPP/88+ndu3emTZuWbt26FawUAABqR90GkDFjxmTSpElJkkMPPXS+fb/1rW+VKAkAAGpe3QaQxx9/vNJueXWrt956K88991ymTJmSddddN2uttVb54gAAoEbV7TkgzzzzTJJZh1itttpqeemll7LHHntk1VVXzWc/+9nssssu6d+/fz75yU/mgQceqHK1AABQG+o2gLz66qtJkhVXXDF//vOfs/nmm+emm27KjBkz5uj3yCOPZLvttsvIkSOrUSYAANSUug0g7733XpJk8uTJ2WuvvfL+++/n61//ep5++ulMnTo1L730Uk444YR06tQpM2bMyGGHHTbHYVu0v759+1Zt3PZaNgC0F99ddBR1G0Def//9JMmkSZMyYcKEnHrqqbn44ouz0UYbpVu3bvn4xz+en/70p/ntb3+bZNZVs/77v/+7miXXnU6d2mfzbM247bVsAGgvvrvoKOp2S21ubq60N9lkk5x00klz7TdkyJBstdVWSZI777wz//znP4vU1xE0NTWlqamp1c8vzBgltOeym5qa5tjGOoKO8r4BMH+lP6NLLM/3Tm2p2wCyzDLLVNr77bdfGhoa5tl37733rrTvv//+dq2rI2lsbExjY2Orn1+YMUpoz2U3NjZ2uA/KjvK+ATB/pT+jSyzP905tqdsAsvzyy1faG2ywwXz7rr/++pX2a6+91m41AQBAravbANK/f/9W92155/OZM2e2RzkAAFAX6jaAbLLJJpX2uHHj5tv3jTfeqLRXX3319ioJAABqXt0GkO22267SHj169Hz7Pvjgg5V2y+ACAAAsnLoNIAMHDsw666yTJLn77rvzyCOPzLXfhAkTcu211yZJBgwYIIAAAMBiqNsAkiQ//vGPk8y6JO8BBxyQV155ZY7p06ZNy8EHH5x33nknSfLtb3+7eI0AAFBLulS7gGo6+OCDc+ONN+aGG27Iiy++mE033TRHHHFEttxyy7z55pu54IIL8txzzyWZdcjWEUccUeWKAQCgY6vrAJIkI0eOzBFHHJErrrgi7777bn75y19+pM8uu+yS66+/fr73CgEAABasrg/BSpKllloqI0aMyJgxY7L//vunX79+6datW1ZbbbV87nOfy/XXX5/Ro0dnueWWq3apAADQ4dX9HpDZdthhh+ywww7VLgMAAGpa3e8BAQAAyhFAAACAYgQQAACgGAEEAAAoRgABAACKEUAAAIBiBBAAAKAYAQQAAChGAAEAAIoRQAAAgGIEEAAAoBgBBAAAKEYAAQAAihFAAACAYgQQAACgGAEEAAAoRgABAACKEUAAAIBiBBAAAKAYAQQAAChGAAEAAIoRQAAAgGIEEAAAoBgBBAAAKEYAAQAAihFAWCx9+/atdgmLpS3q7+ivQUu1tC5A/fJZBks2AYTF0qlTx96E2qL+jv4atFRL6wLUL59lsGTzF0qbaGpqSnNzc7XLWGQLqr8167ekvAZLSh0A1dbU1JSmpqaaWvaijrukfze0rG9Jr5XFJ4DQJhobG6v2Id8WFlR/a9ZvSXkNlpQ6AKqtsbExjY2NNbXsRR13Sf9uaFnfkl4ri08AAQAAihFAAACAYgQQAACgGAEEAAAoRgABAACKEUAAAIBiBBAAAKAYAQQAAChGAAEAAIoRQAAAgGIEEAAAoBgBBAAAKEYAAQAAihFAAACAYgQQAACgGAEEAAAoRgABAACKEUAAAIBiBBAAAKAYAQQAAChGAAEAAIoRQAAAgGIEEAAAoBgBBAAAKEYAAQAAihFA5mHixIlZffXV09DQkMMOO6za5QAAQE0QQObh+OOPz+uvv17tMgAAoKYIIHNx6623ZsSIEdUuAwAAao4A8iETJ07M4MGDq10GAADUJAHkQ2YfetWzZ89qlwIAADVHAGmh5aFXZ511VpWrAQCA2iOA/EfLQ68GDx6cnXbaqcoVAQBA7RFA/mP2oVf9+vXLL37xi2qXAwAANUkAyZyHXl1wwQVZfvnlq1wRAADUproPIC0PvTrkkEPyhS98ocoVAQBA7ar7ADL70KtVVlklv/rVr6pdDgAA1LS6DiAtD70aPnx4VlxxxSpXBHPXt2/fapcAwBLCdwIdXd0GkJaHXu27777ZY489qlsQzEenTnX7pwrAh/hOoKOr2y149qFXvXv3zrnnnlvtcqBVmpqa0tTU1KbjNTc3t3nf+Y3RlvUvKUqvV62+jrPV+vq1RjVfA9vzgrVFzW01xuJ+LpfQss6O+H7T9uoygLQ89Orcc89Nnz59qlwRtE5jY2MaGxvbdLzWfhEsTN/5jdGW9S8pSq9Xrb6Os9X6+rVGNV8D2/OCtUXNbTVGR/gx37LOjvh+0/bqMoD8/ve/r7T333//NDQ0fORf//79K31GjBhReX7YsGFVqBgAAGpDXQYQAACgOrpUu4BqOPbYYxd40vmbb76ZIUOGJEl22GGHHHvssUmSDTbYoL3LAwCAmlWXAWTLLbfMlltuOd8+48aNq7TXWGMNV8kCAIA24BAsAACgGAEEAAAoRgABAACKEUAAAIBi6vIk9NZYa621OsTdRQEAoCOxBwQAAChGAAEAAIoRQAAAgGIEEAAAoBgBBAAAKEYAAQAAihFAAACAYgQQAACgGAEEAAAoRgABAACKEUAAAIBiBBAAAKAYAQQAAChGAAEAAIoRQAAAgGIEEAAAoBgBBAAAKEYAAQAAihFAAACAYgQQAACgGAEEAAAoRgABAACKEUAAAIBiBBAAAKAYAQQAAChGAAEAAIoRQAAAqqBv377VLqHNLMy61NJ6s2gEEACAKujUqXZ+hi3MutTSerNobAEslKampjQ3N1e7DD5kYd6X9urbnpqamtLU1FTtMurWvF7/tnxfqrmtdfTta1HrX9j5FvQedZTXseV6LCk1f7imatW3MMub1/ZQ+jtmSXkPWTgCCAulsbHRH/oSaGHel/bq254aGxvT2NhY7TLq1rxe/7Z8X6q5rXX07WtR61/Y+Rb0HnWU17HleiwpNX+4pmrVtzDLm9f2UPo7Zkl5D1k4AggAAFCMAAIAABQjgAAAAMUIIAAAQDECCAAAUIwAAgAAFCOAAAAAxQggAABAMQIIAABQjAACAAAUI4AAAADFCCAAAEAxAggAAFCMAAIAABQjgAAAAMUIIAAAQDECCAAAUIwAAgAAFCOAAAAAxQggAABAMQIIAABQjAACAAAUI4AAAADFCCAAAEAxAsh/vPPOOznzzDOz9dZbp1evXunatWv69u2bnXfeORdccEE++OCDapcIAAAdXpdqF7AkGDNmTPbbb7+MHz9+jufHjx+fu+++O3fffXfOPffc3Hzzzenfv3+VqgQAgI6v7gPIE088kUGDBmXy5MlJkl122SV77LFHevXqlZdffjkjRozIM888k2eeeSa77LJLHnnkkfTs2bO6RQMAQAdV94dgHXvssZXwMXz48Pzxj3/MkUcemX322Sff//738/jjj+fAAw9Mkrz44os57bTTqlkuAAB0aHUdQJ577rn8+c9/TpLsscce+eY3v/mRPl26dMnFF1+c1VZbLUly2WWXpampqWidAABQK+o6gNx1112V9sEHHzzPft27d8/uu++eZNbJ6i+88EK71wYAALWorgNI586ds9FGG2X55ZfPeuutN9++K620UqX9zjvvtHdpAABQk+o6gBx55JF5+umn8+6772bjjTeeb99nn3220u7Vq1d7lwYAADWprgNIa73++uu5/fbbkyS9e/fOOuusU+WKAACgYxJAWuG73/1u5UaE++23Xzp18rIBAMCi8Et6AS644IKMHDkySbLsssvmBz/4QZUrAgCAjksAmY8bb7wxRx11VOXx8OHDK5fjrRd9+/atiWWw5KjV93tu61Wr6wodyYL+Dtvz73RJ/gyYV21Lcs3UDgFkHm644Ybss88+lXt+HHfccTnooIOqXFV5JQ43c0hbfanV93tu61Wr6wodyYL+Dtvz73RJ/gyYV21Lcs3UDlvZXFxyySXZd999M3369CTJoYcemrPPPrvKVdW+pqamNDc3V7sMCmn5fjc1NX3kBp8Luz205fYzt3raY7y2Xk57j1sL2mI7ac/3zWdg9Xz4M2lu7Vo0r/Wr9fWmugSQDznllFNyxBFHZMaMGUmSr33ta7n00kvT0NBQ5cpqX2Njox9NdaTl+93Y2JjGxsZ5Tl/Y8dqitg/X0x7jtfVy2nvcWtAW20l7vm8+A6vnw59Jc2vXonmtX62vN9XVpdoFLClmzpyZI488MhdeeGHlueOOOy5nn3228AEAAG3EHpDMCh+HHHLIHOHjtNNOy69+9SvhAwAA2pA9IEm+9a1v5eqrr04y6+Sr8847L9/85jerXBUAANSeug8gl19+eS644IIks8LH5ZdfnoMPPrjKVQEAQG2q60Ow3njjjRx33HGVx2eddZbwAQAA7aiu94D86le/yqRJk5Ika6yxRvr3758bb7xxgfNtueWWWWONNdq5OgAAqD11HUBGjBhRab/yyivZc889WzXfZZddlsMOO6ydqgIAgNpVt4dgvfXWW66TDwAAhdXtHpDevXu7wycAABRWt3tAAACA8gQQAACgGAEEAAAoRgABAACKEUAAAIBiBBAAAKAYAQQAAChGAAEAAIoRQAAAgGIEEAAAoBgBBAAAKEYAAQAAihFAAACAYgQQAACgGAEEAAAoRgABAACKEUAAAIBiBBAAAKAYAQQAAChGAAEAAIoRQAAAgGIEEAAAoBgBBAAAKEYAAQAAihFAAACAYgQQAACgGAGEJEnfvn2X6PFY8tT6NjOvepa0Ouel1PvTUV6PBeno67Gk/D2Wnq+aOmLNHZ3XvHYIICRJOnVq202hrcdjyVPr28y86lnS6pyXUu9PR3k9FqSjr8eS8vdYer5q6og1d3Re89rhnWQOTU1NaW5ubtfxFnUZbV1bPWnP167l2G35fi/sslvTt6mpaaHHmNf0jrI9zmu9F7Vva16v1i5vYS1o7LZY9ry254XZTua3rX34+fb4zG3L178tPq/b4u90Sbao20lHtKTUv6TUwaITQJhDY2Njm374z228RV1GW9dWT9rztWs5dlu+3wu77Nb0bWxsXOgx5jW9o2yP81rvRe3bmtertctbWAsauy2WPa/teWG2k/ltax9+vj0+c9vy9W+Lz+u2+Dtdki3qdtIRLSn1Lyl1sOgEEAAAoBgBBAAAKEYAAQAAihFAAACAYgQQAACgGAEEAAAoRgABAACKEUAAAIBiBBAAAKAYAQQAAChGAAEAAIoRQAAAgGIEEAAAoBgBBAAAKEYAAQAAihFAAACAYgQQAACgGAEEAAAoRgABAACKEUAAAIBiBBAAAKAYAQQAAChGAAEAAIoRQAAAgGIEEAAAoBgBJElzc3Ouueaa7LzzzunVq1e6d++e/v3759BDD80jjzxS7fIAAKBmdKl2AdU2ZcqU7L333rn11lvneH7cuHEZN25crr766pxxxhk54YQTqlQhAADUjroPIEcccUQlfAwYMCCDBw/OyiuvnL/+9a85//zz8/777+fEE0/M6quvnoMOOqjK1QIAQMdW1wHkzjvvzDXXXJMk2XHHHXPrrbeme/fuSZL9998/X/va1/LZz342EyZMyPHHH58vf/nLWW655apZMgAAdGh1fQ7IWWedlSTp0qVLLr744kr4mG3AgAE599xzkyQTJkzIRRddVLxGAACoJXUbQN5+++3cddddSZLPf/7z6d+//1z77bvvvunbt2+S5Pe//32x+gAAoBbVbQC57777MnPmzCTJTjvtNM9+nTp1ynbbbZck+ctf/pJ33nmnSH0AAFCL6jaAPP3005X2xhtvPN++G264YZJZl+t96qmn2rUuAACoZXUbQMaNG1dpr7XWWvPt269fv0r75ZdfbqeKAACg9jU0Nzc3V7uIathjjz1y0003JUneeeed9OzZc559//d//zd77bVXkuTss8/O8ccf3+b1/PnPf843vvGNyuMf//jHWX/99dt8OXMzfvz4rLTSSuncufM82wvTt63HsOw5pydJnz59WvUe1vt6f7jvgsZoz/We17Ln9vyCLGi+1ix7YcZd2PVe1HVa3NerNctuq/f7w8tZ1PoXZhmtWafS73d7zLeo690RP9faYltr6/mWlPVemDFau820hX/84x859dRTK48vuuiibLvttkWWXUvq9jK8kydPrrQ/fPWrD2s5veV8bemf//znHI9bbtwAACx5Pvz7jdap20OwZsyYUWl369Ztvn1bTm85HwAAsHDqNoD06NGj0v7ggw/m23fatGmVdteuXdutJgAAqHV1ewjWsssuW2lPnTp1vntBpk6dWmm3DC5taccdd5zj8RprrNFuywIAYOFNmTIlr7zySuXxh3+/0Tp1G0BannT+9ttvZ4UVVphn37fffrvSbq+TnFZdddUceOCB7TI2AAAsKer2EKz11luv0m6ZZOem5QlGa665ZrvVBAAAta5uA8jsmwsmyTPPPDPfvrOnNzQ0ZKONNmrXugAAoJbVbQD59Kc/naWWWipJcs8998yzX1NTU+69994kyWabbTbf+4UAAADzV7cBpGfPntlpp52SJKNGjZrnYVgjR46s3ORmn332KVYfAADUorq9E3qS3HHHHdl1112TJJ/5zGdy++23Z7nllqtMf/bZZ7PttttmwoQJWW655fJ///d/6d27d7XKBQCADq+uA0iSfPWrX80f/vCHJEn//v1z5JFHpl+/fvnb3/6W4cOH57333kuSnH/++RkyZEg1SwUAgA6v7gPI5MmTs+eee+aOO+6Y6/SGhoYMHTo0Q4cOLVwZAADUnroPIEnS3Nyca665JiNGjMjjjz+eiRMnpnfv3tlmm21y3HHHZeutt652iQAAUBMEEAAAoJi6vQoWAABQngACAAAUI4AAAADFCCAAAEAxAggAAFCMAAIAABQjgAAAAMUIIAAAQDECCAAAUIwAUmeam5tzzTXXZOedd06vXr3SvXv39O/fP4ceemgeeeSRapdHB9fe21f//v3T0NDQqn//+te/2mCNYJajjjoqDQ0NGTZsWLVLoQa15fa1ww47tPpz8qGHHlr84mERdKl2AZQzZcqU7L333rn11lvneH7cuHEZN25crr766pxxxhk54YQTqlQhHVl7b1+TJk3Kyy+/3BalwkIZM2ZMzj///GqXQY1q6+3rqaeearOxoL0IIHXkiCOOqPw4HDBgQAYPHpyVV145f/3rX3P++efn/fffz4knnpjVV189Bx10UJWrpaNp7+3rqaeeSnNzc5LkRz/6UQYOHDjf/iuuuOLCrwR8yGOPPZY999wzM2fOrHYp1KC23r5ee+21TJgwIUnyjW98I7vtttt8+6+//vptslxYaM3UhTvuuKM5SXOS5h133LF5ypQpc0x/9tlnm3v16tWcpLlXr17NkyZNqlKldEQltq/f/va3lWX84x//aKvSYZ5uu+225p49e1a2uyTNQ4cOrXZZ1Ij22L5uu+22ylh//OMf26ZQaAfOAakTZ511VpKkS5cuufjii9O9e/c5pg8YMCDnnntukmTChAm56KKLitdIx1Vi+3ryySeTJN27d88666yzmBXDvE2bNi1Dhw7N7rvvnokTJ1a7HGpMe25fsz8nk2TjjTdu07GhLQkgdeDtt9/OXXfdlST5/Oc/n/79+8+137777pu+ffsmSX7/+98Xq4+OrdT2NfuLdcMNN0ynTj66aB933XVXBgwYkFNPPTUzZ87MMsssk+985zvVLosa0d7b1+zPyZVWWimrrbZam40Lbc23eB247777KseX7rTTTvPs16lTp2y33XZJkr/85S955513itRHx1Zi+2pubs7TTz+dJNlkk00Wo1qYv6uuuipjx45NkgwcODAPP/xwBg0aVOWqqBXtvX3NDiA+J1nSCSB1YPYPt2TBu2Q33HDDJLN+8LmSBq1RYvt6+eWXM2nSpDmWMWnSpNx3330ZPXp0Hn/88TQ1NS1s6TBXffr0yXnnnZeHHnqoss1CW2mv7Wv69On5xz/+keT/fU5Onjw5Dz74YEaPHp1HHnkkH3zwQZstDxaHq2DVgXHjxlXaa6211nz79uvXr9J2yVNao8T21fK45k6dOmWvvfbKqFGj5ggdvXr1ytFHH50TTjghPXr0aPXY0NLRRx+d4cOH24ZoF+25ff3973/P9OnTkyQrrLBCDj/88IwcOTJTp06t9FluueVy+OGHZ9iwYa4USFUJIHVg/PjxlXbv3r3n23ellVaqtGdfyg/mp8T21TKAfPe7351rnwkTJuSUU07J6NGjc+utty6wFpibBV3eGRZHe25fLT8nf/KTn8y1z3vvvZdzzjknt912W/74xz/m4x//eLvVA/PjEKw6MHny5Er7w1cn+rCW01vOB/NSYvtq+cXatWvXnHjiifn73/+eqVOn5vXXX89FF12UVVddNUny8MMPZ99993XfBqCutPycbGhoyJAhQ/K3v/0tkydPzvjx4zNy5Mist956SZIXX3wxgwYN8j1P1QggdWDGjBmVdrdu3ebbt+X0lvPBvJTYvl577bUkswLM3XffnTPPPDMbbLBBunXrllVXXTVHHHFEHnnkkXzsYx9LMuvOwtddd93CrAZAh/bqq68mmRU+rrvuupx//vnZfPPN06NHj/Tu3Tv77rtvHn744Wy++eZJkmeffTa//vWvq1gx9UwAqQMtjzVd0Alo06ZNq7S7du3abjVRO0psX/fff38mTZqU559/Pttss81c+6y++uo555xzKo/dywaoJ9dcc03+/e9/56WXXsree+891z4rrLBCLr300spjn5NUiwBSB5ZddtlKu+XJaHPTcrqTMGmNUtvXcsstN8dJ7HPzpS99Kcsvv3yS5IEHHnAYFlBXll566Xnei2m2LbbYIhtttFGSZOzYsZU9zFCSAFIHevbsWWm//fbb8+3bcnqfPn3aqyRqyJK0fXXu3Dnrrrtukll7W1xIAeCjNthgg0r79ddfr2Il1CsBpA7MPuksSV555ZX59v3nP/9Zaa+55prtVhO1Y0nbvpZaaql2GRegVvicpNpchrcOtLzR0TPPPFO5G/XcPPPMM0lmncQ2exctzE97b19jx47NqFGj8sYbb2TrrbfOF7/4xfn2n304QdeuXee47C9ArXrrrbcycuTIvPHGG1lvvfVy8MEHz7d/y8OuZl9BEEqyB6QOfPrTn678b8c999wzz35NTU259957kySbbbbZHIfWwLy09/b1+uuv5/jjj8+ZZ56ZCy+8cL59n3/++cpelv/6r/9K586dW7UMgI5s2rRpOeaYY3L66afnrLPOmm/fd999N48++miSWRfvmH31QChJAKkDPXv2zE477ZQkGTVq1DwPkxk5cmTlpnL77LNPsfro2Np7+/rEJz5RObH89ttvn+9hXqeffnqlvaD/AQSoFauvvnrlcNgnn3wyDz300Dz7/vKXv8yUKVOS+JykegSQOvHtb387yazLpO6///5577335pj+7LPP5rjjjksy62pD3/jGN4rXSMfVnttX9+7dK/0/+OCDHHDAAZk0adJH+p111lm58sorkyTrrrtuDj300EVaF4CO6Jhjjqm0Dz300LmeXD5y5MiceeaZSZJevXpVPruhNOeA1IlddtklX/nKV/KHP/whDzzwQDbbbLMceeSR6devX/72t79l+PDhlR+Nv/jFL9K7d+8qV0xHsrjb15/+9KfssMMOSZLtttsuf/rTn+aYfvLJJ+eWW27JP/7xj9x///3ZaKONMmTIkKy33noZP358rrvuutx3331JZgWca665ZoF3ZQfoSMaNG1e5xO6aa66ZcePGzTH9m9/8Zn73u9/lvvvuy/PPP59NN900gwcPzuabb5533303o0aNyi233JIk6dKlSy6//PL07du39GpAEgGkrlxxxRV57733cscdd2Ts2LH57//+7zmmNzQ0ZOjQoRkyZEiVKqQja8/ta4UVVsidd96ZvfbaK48++mheffXVnHzyyR/p169fv1x77bUZOHDgIq8HQEfUpUuXjBo1Kvvtt1/++Mc/ZsKECZW9HS2ttNJKueSSS7L77rtXoUqYxSFYdWTppZfO7bffnquuuiqf+9zn0qdPnyy11FJZddVVs/fee+e+++7L0KFDq10mHVR7b1/9+vXLgw8+mCuuuCJf+MIX0rdv3yy11FLp3bt3PvOZz+Sss87Ks88+m6233roN1wqg4+jZs2dGjx6dG2+8MXvuuWdWW221LLXUUllxxRXziU98Iqecckr+/ve/Z4899qh2qdS5hubm5uZqFwEAANQHe0AAAIBiBBAAAKAYAQQAAChGAAEAAIoRQAAAgGIEEAAAoBgBBAAAKEYAAQAAihFAAACAYgQQAACgGAEEAAAoRgABAACK6VLtAgAAoNrWX3/9BfZZffXVM2bMmALV1DYBBACAunf00UfPc9qoUaPyyiuv5NOf/nTBimpXQ3Nzc3O1iwAAgCXRn//85wwePDgbb7xxrrnmmnTt2rXaJXV4AggAAMzFu+++m89//vOZPHlyRo0alTXXXLPaJdUEJ6EDAMBcnHPOOXn77bdz1FFHCR9tyB4QAAD4kJdeeimDBg3KKquskttvv92hV23IHhAAAPiQiy++OE1NTRkyZIjw0cbsAQEAgBbeeOON7LTTTunVq1fuvPNOAaSN2QMCAAAt3HTTTZk+fXr22Wcf4aMdCCAAANDCH//4xyTJ7rvvXuVKapMAAgAA//HGG2/k6aefzsYbb+zKV+1EAAEAgP947LHHkiRbbbVVlSupXQIIAAD8x1NPPZUk2XTTTatcSe0SQAAAWCKdcsopWX/99XPuuee2qn9zc3NuvvnmHHbYYdlqq62yySabZMcdd8wJJ5yQJ598slVjvPzyy0mSVVZZZZHrZv66VLsAAAD4sAcffDAjR45sdf+pU6fmuOOOy5/+9Kc5nn/ttdfy2muv5eabb87xxx+fwYMHz3ect99+O0my3HLLLXTNtI4AAgDAEuXpp5/O0UcfnZkzZ7Z6nh/96EeV8LH22mtnn332Se/evfPMM89k5MiRmTx5cn75y19m5ZVXzpe//OV5jrMwoYdFI4AAALDEuPfee/O9730v77//fqvnuf/++3PLLbckST71qU/lwgsvTLdu3ZLMupTuV7/61RxwwAGZOHFifvKTn2SnnXbKsssu2y71s2DOAQEAYLF88MEHC+wzY8aMBY5xzjnn5Jvf/GYmTZq0UMu/9NJLkyRdunTJ6aefXgkfs6299to5+eSTkyQTJ07M73//+4Uan7YlgAAAsMiuu+667L///vPdYzFu3Lh88YtfzBNPPDHX6Q888EC+8IUv5LzzzsvMmTOz9NJL5/DDD2/V8idOnJgHHnggSfLZz342/fr1m2u/3XbbLb169UqS3H777a0am/YhgAAAsEjGjBmTYcOG5emnn87gwYMzZcqUj/R5/fXXc/jhh2fcuHH5xje+kVdfffUjfUaNGlV5fuONN87vf//77LDDDq2q4dFHH62cK/KpT31qnv06deqU//qv/0qSPPHEE3n33XdbNT5tTwABAGCRfOITn8h6662XZNYN/L71rW/NcTjW+PHjc9hhh+X1119Pkmy33XZZffXV5zrWSiutlB//+Mf53e9+l3XWWafVNbzwwguV9uxa5mX2uM3NzXn++edbvQzalgACAMAiWWGFFXLppZemf//+SWYdSnXcccdlxowZeeedd3L44YdX7qvxuc99LmeeeWYaGho+Ms6BBx6Ye+65JwceeGA6d+68UDW89tprlfa8ws1sLe/t0XI+yhJAAABYZL169cqIESPysY99LMmsw7K++93v5utf/3pl78Q222yT//mf/0mXLnO/AOsmm2yS7t27L9LyZ9+3I0lWXHHF+fbt2bNnpT1x4sRFWh6LTwABAGCxrLzyyrn88suz8sorJ5l1kvczzzyTJBk4cGB+85vfpGvXru2y7KlTp1baH7761Ye1rKHlfJQlgAAAsNj69euXCy+8MJ06/b+fl717984FF1yQHj16tNtyW17ed0Ehp+X0BV0WmPYjgAAAsNiampoql9Gd7a233soll1zSrstteejW9OnT59u35Qny7bVHhgUTQAAAWCwzZ87MiSeemDvuuCPJrL0hSy+9dJLkt7/9bS666KJ2W/bs5STJtGnT5tu3ZQBZ0OFatB8BBACAxTJs2LCMGjUqSbLaaqvliiuuyPDhwys/8s8666xcffXV7bLs5ZdfvtJe0InlLaevtNJK7VIPCyaAAACwyH7605/muuuuS5L06dMnl19+eVZbbbV86lOfyrnnnpullloqSXLaaaflhhtuaPPlr7XWWpV2Y2PjfPv+61//qrRXW221Nq+F1hFAAABYJLfccksuu+yyJLMucXvZZZdlzTXXrEzfbrvtctZZZ6Vz585pbm7OSSedlJdeeqlNa1h77bUr7ZY3JZyb2dMbGhqy7rrrtmkdtN7/DylQxfkA1I++AAAAAElFTkSuQmCC",
      "text/plain": [
       "<Figure size 400x400 with 1 Axes>"
      ]
     },
     "metadata": {
      "image/png": {
       "height": 400,
       "width": 400
      }
     },
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "n2 max count 64.0\n",
      "n2 max bin [6581870.875, 6669156.42]\n",
      "\n",
      "cb max count 15.0\n",
      "cb max bin [2997721.4499999997, 3083021.12]\n"
     ]
    }
   ],
   "source": [
    "spec_family = 'hAT'\n",
    "chromosome = 'Chr IV' \n",
    "\n",
    "peak_tracker_list = track_down_hist_peaks(N2_all_df, CB_all_df, spec_family, chromosome)\n",
    "\n",
    "n2_values = peak_tracker_list[0].tolist()\n",
    "n2_bins = peak_tracker_list[1].tolist()\n",
    "\n",
    "n2_max_value = (max(n2_values))\n",
    "n2_max_value_index = (n2_values.index(n2_max_value))\n",
    "\n",
    "n2_max_value_bin = n2_bins[n2_max_value_index-1:n2_max_value_index+1]\n",
    "print('n2 max count', n2_max_value)\n",
    "print('n2 max bin', n2_max_value_bin)\n",
    "\n",
    "cb_values = peak_tracker_list[3].tolist()\n",
    "cb_bins = peak_tracker_list[4].tolist()\n",
    "\n",
    "cb_max_value = (max(cb_values))\n",
    "cb_max_value_index = (cb_values.index(cb_max_value))\n",
    "\n",
    "cb_max_value_bin = cb_bins[cb_max_value_index-1:cb_max_value_index+1]\n",
    "print()\n",
    "print('cb max count', cb_max_value)\n",
    "print('cb max bin', cb_max_value_bin)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 6.4.4. <a id='toc6_4_4_'></a>[FX: `make_every_fam_grid`](#toc0_)\n",
    "- make plot of every single TE family distribution by chr in a huge grid\n",
    "- columns: chromosome, rows: TE family\n",
    "- x-axis: relative start position\n",
    "- y-axis: count\n",
    "- this is not a great looking graph, but can be helpful for getting an overview of the data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 74,
   "id": "de8c0e8d",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2022-07-26T17:56:47.410493Z",
     "start_time": "2022-07-26T17:56:07.621892Z"
    },
    "code_folding": []
   },
   "outputs": [],
   "source": [
    "#N2 86abb1\n",
    "#CB b0948f\n",
    "# ,facet_kws=dict(sharex='col')\n",
    "def make_every_fam_grid(df, genome: str, numbins: int, sharey: bool = False, to_save: bool = False):\n",
    "    \"\"\"  \n",
    "    df: TE dataframe\n",
    "    genome: str, name of the genome that will be inserted into the title\n",
    "    numbins: int, number of equally sized bins the data will be divided into\n",
    "    sharey: bool, whether or not the y-axis should be shared each individual histogram in the entire grid\n",
    "    to_save: bool, whether or not to save the graph as a pdf (if False, graph will only be output into the notebook)\n",
    "\n",
    "    \"\"\"\n",
    "\n",
    "    color = colors[genome]\n",
    "    \n",
    "    if sharey:\n",
    "        shareystr = 'nosharey'\n",
    "    else: \n",
    "        shareystr = 'sharey'\n",
    "        \n",
    "    g = sns.displot(data=df, x=\"Relative Start Position\", col=\"Chromosome\", row = 'Family', bins = numbins, kind=\"hist\", color = color, edgecolor = color, facet_kws=dict(sharey=sharey))\n",
    "    \n",
    "    g.set_axis_labels(\"Relative Start Position\", \"Count\")\n",
    "    \n",
    "    if to_save == True:  \n",
    "        plt.savefig(f'{genome}_ultimate_byfamandchr_hist_{numbins}_LTRRet_{shareystr}.pdf', dpi=300, facecolor = 'w', edgecolor = 'w', format='pdf')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 73,
   "id": "e29ee811",
   "metadata": {
    "scrolled": false
   },
   "outputs": [],
   "source": [
    "# make_every_fam_grid(N2_all_filter_df, 'N2', 100, sharey = False, to_save = False)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 6.5. <a id='toc6_5_'></a>[TE COUNTS BAR GRAPH](#toc0_)\n",
    "### 6.5.1. <a id='toc6_5_1_'></a>[FX: `make_te_counts_perfam`](#toc0_)\n",
    "- creates a bar graph of the number of DNA and Retro elements (DNA in one graph, Retro in the second) in a given genome"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 78,
   "id": "dde0a270",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2022-11-28T22:06:20.030793Z",
     "start_time": "2022-11-28T22:06:20.023910Z"
    }
   },
   "outputs": [],
   "source": [
    "def make_te_counts_perfam(strain: str, dna_tes, retro_tes, to_save: bool = False):\n",
    "    \"\"\" \n",
    "    strain: str, name of the strain that will be used in the graph title and file name\n",
    "    dna_tes: dna TEs dataframe\n",
    "    retro_tes: retro TEs dataframe \n",
    "    to_save: bool, whether or not to save the graph as a pdf (if False, graph will only be output into the notebook)\n",
    "    \"\"\"\n",
    "    \n",
    "    sns.set(style=\"white\")\n",
    "        \n",
    "    sns.set_palette(sns.color_palette(dna_color_list))\n",
    "\n",
    "    fig, ax = plt.subplots(figsize=(12, 8))\n",
    "    ax = sns.countplot(data = dna_tes, x = 'Family', order = dna_order)\n",
    "    ax.set_ylabel(\"Count\")\n",
    "    for i in ax.containers:\n",
    "        ax.bar_label(i,)\n",
    "    plt.title(f'{strain} DNA TE counts per family')                                   \n",
    "\n",
    "    if to_save == True:\n",
    "    \n",
    "        plt.savefig(f'{strain}_DNA_counts_all_LTRRet.pdf', dpi=300, facecolor = 'w', edgecolor = 'w', format='pdf')\n",
    "    \n",
    "    sns.set_palette(sns.color_palette(retro_color_list))\n",
    "\n",
    "    fig, ax2 = plt.subplots(figsize=(12, 8))\n",
    "    ax2 = sns.countplot(data = retro_tes, x = 'Family',  order = retro_order)\n",
    "    ax2.set_ylabel(\"Count\")\n",
    "    for i in ax2.containers:\n",
    "        ax2.bar_label(i,)\n",
    "    plt.title(f'{strain} Retro-TE counts per family')                                   \n",
    "    \n",
    "    if to_save == True:\n",
    "        \n",
    "        plt.savefig(f'TEUlt_{strain}_retro_counts_all_LTRRet.pdf', dpi=300, facecolor = 'w', edgecolor = 'w', format='pdf')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 83,
   "metadata": {},
   "outputs": [],
   "source": [
    "# make_te_counts_perfam('N2', N2_dna_tes, N2_retro_tes, False)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "below: runs the `make_te_counts_perfam` function on all genomes in the `strain_dict` list defined in the [graphing variables](#61-graphing-variables) cell"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 76,
   "id": "852cd5eb",
   "metadata": {},
   "outputs": [],
   "source": [
    "# for key, value in strain_dict.items():\n",
    "#     dna_df = value[0]\n",
    "#     retro_df = value[1]\n",
    "#     make_te_counts_perfam(key, dna_df, retro_df)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 6.5.2. <a id='toc6_5_2_'></a>[FX: `make_class_counts_allgenomes`](#toc0_)\n",
    "- creates a paired bar graph style plot of the counts of either DNA or Retro elements for all genomes "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 86,
   "id": "7800e534",
   "metadata": {},
   "outputs": [],
   "source": [
    "def make_class_counts_allgenomes(te_class: str, to_save: bool = False):\n",
    "    \"\"\" \n",
    "    te_class: str, one of either 'retro' or 'dna'\n",
    "    to_save: bool, whether or not to save the graph as a pdf (if False, graph will only be output into the notebook)\n",
    "    \"\"\"\n",
    "\n",
    "    if te_class == 'retro':\n",
    "        df = all_retro_tes\n",
    "        title = 'Retro-'\n",
    "        filestr = 'retro'\n",
    "        order = retro_order\n",
    "        \n",
    "    elif te_class == 'dna':\n",
    "        df = all_dna_tes\n",
    "        title = 'DNA '\n",
    "        filestr = 'dna'\n",
    "        order = dna_order\n",
    "    \n",
    "    sns.set(style=\"white\")\n",
    "    \n",
    "    fig, ax = plt.subplots(figsize=(12, 8))\n",
    "    ax = sns.countplot(data = df, x = 'Family', order = order, hue='Strain', palette=colors)\n",
    "    ax.set_ylabel(\"Count\")\n",
    "    for i in ax.containers:\n",
    "        ax.bar_label(i, fontsize=7, label_type='edge')\n",
    "    plt.title(f'{title}TE counts')   \n",
    "    \n",
    "    if to_save == True: \n",
    "        plt.savefig(f'{filestr}counts_allgenomes.pdf', dpi=300, facecolor = 'w', edgecolor = 'w', format='pdf')\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 88,
   "id": "a54f0b94",
   "metadata": {},
   "outputs": [],
   "source": [
    "# make_class_counts_allgenomes('dna', False)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 6.6. <a id='toc6_6_'></a>[TE SIZES GRAPHS](#toc0_)\n",
    "### 6.6.1. <a id='toc6_6_1_'></a>[boxplot of all sizes on one plot](#toc0_)\n",
    "- this is not a great graph but can be helpful for getting an overview of the data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "d1bd8d8a",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2022-08-07T07:01:15.369070Z",
     "start_time": "2022-08-07T07:01:15.088780Z"
    }
   },
   "outputs": [],
   "source": [
    "# sns.boxplot(x = 'Family', y = 'Size', data=N2_dna_tes)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 6.6.2. <a id='toc6_6_2_'></a>[FX: `broken_axis_box` ](#toc0_)\n",
    "- creates a broken axis boxplot of the TE sizes in either the DNA or Retro class\n",
    "- used these links to make this fx:\n",
    "    - https://gist.github.com/pfandzelter/0ae861f0dee1fb4fd1d11344e3f85c9e \n",
    "    - https://matplotlib.org/2.0.2/examples/pylab_examples/broken_axis.html"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 96,
   "id": "f8afb89c",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2022-08-07T06:59:47.814814Z",
     "start_time": "2022-08-07T06:59:47.130610Z"
    }
   },
   "outputs": [],
   "source": [
    "dna_order = ['Zator', 'hAT', 'CMC', 'Tc1-Mariner', 'Sola', 'Novosib', 'Helitron', 'MITE']\n",
    "retro_order = ['Gypsy', 'Copia', 'LINE', 'SINE', 'ERV']\n",
    "\n",
    "def broken_axis_box(data, genome: str, break_num: int, end_num: int, title: str, filename: str, order: list, showmeans: bool, to_save: bool = False):\n",
    "    \"\"\" \n",
    "    data: TE dataframe\n",
    "    genome: str, used in the title and filename\n",
    "    break_num: int, number to break the axis at\n",
    "    end_num: int, number to cap the y-axis at (can think of this as the y-limit)\n",
    "    title: str, something like 'DNA ' or 'Retro' to indicate what class of TEs are being plotted\n",
    "    filename: str, something like 'DNA ' or 'Retro' to indicate what class of TEs are being plotted\n",
    "    order: list, what order to plot the families along the x-axis\n",
    "    showmeans: bool, whether or not to include a red dot at the mean TE size of each family\n",
    "    to_save: bool, whether or not to save the graph as a pdf (if False, graph will only be output into the notebook)\n",
    "    \"\"\"\n",
    "\n",
    "    sns.set(style=\"whitegrid\")\n",
    "    pal = sns.color_palette('colorblind')\n",
    "\n",
    "    f, (ax1, ax2) = plt.subplots(ncols=1, nrows=2, sharex=True, figsize = (12, 8))\n",
    "    flierprops = dict(marker='o', fillstyle = 'none')\n",
    "    meanpointprops = dict(marker='D', markerfacecolor='red', markeredgecolor = 'black')\n",
    "\n",
    "\n",
    "    ax1 = sns.boxplot(x = 'Family', y = 'Size', data=data, palette=pal, ax=ax1, order = order, fliersize = 5, flierprops = flierprops, showmeans = showmeans, meanprops=meanpointprops)\n",
    "    ax2 = sns.boxplot(x = 'Family', y = 'Size', data=data, palette=pal, ax=ax2, order = order, fliersize = 5, flierprops = flierprops, showmeans = showmeans, meanprops=meanpointprops)\n",
    "\n",
    "    ax1.set_ylim(break_num, end_num) # set y limits for both, if you want the scale to stay the same you have to have the same y range for both plots/axes\n",
    "    ax2.set_ylim(0, break_num)\n",
    "\n",
    "    ax1.get_xaxis().set_visible(False) # get rid of x axis for upper plot (it shares with the bottom one)\n",
    "\n",
    "    ax1.set_ylabel(\"\") # get rid of y labels for both\n",
    "    ax2.set_ylabel(\"\")\n",
    "    f.text(0.05, 0.55, \"Size of TE (bp)\", va=\"center\", rotation=\"vertical\") # make ur own y label (diy)\n",
    "\n",
    "    # ax1.get_legend().remove() # get rid of automatic legends\n",
    "    # ax2.get_legend().remove()\n",
    "    # ax2.legend(loc=(1.025, 0.5), title=\"Design\") # make ur own legend\n",
    "\n",
    "\n",
    "    ax1.xaxis.tick_top()\n",
    "    ax2.xaxis.tick_bottom()\n",
    "\n",
    "    f.subplots_adjust(left=0.15, right=0.85, bottom=0.15, top=0.85, hspace=0.1)\n",
    "\n",
    "    d = .01  # how big to make the diagonal lines in axes coordinates\n",
    "    # arguments to pass to plot, just so we don't keep repeating them\n",
    "    kwargs = dict(transform=ax1.transAxes, color=\"k\", clip_on=False)\n",
    "    ax1.plot((-d, +d), (-d, +d), **kwargs)        # top-left diagonal\n",
    "    ax1.plot((1 - d, 1 + d), (-d, +d), **kwargs)  # top-right diagonal\n",
    "\n",
    "    kwargs.update(transform=ax2.transAxes)  # switch to the bottom axes\n",
    "    ax2.plot((-d, +d), (1 - d, 1 + d), **kwargs)  # bottom-left diagonal\n",
    "    ax2.plot((1 - d, 1 + d), (1 - d, 1 + d), **kwargs)  # bottom-right diagonal\n",
    "\n",
    "    ax1.title.set_text(f'Size Distributions of {genome} {title}TEs')\n",
    "    # plt.legend()\n",
    "\n",
    "    if to_save == True:\n",
    "        plt.savefig(f'TEUlt_{genome}_{filename}_size_dist_broken_LTRRet.png', dpi=300, facecolor = 'w', edgecolor = 'w', format='png')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 92,
   "metadata": {},
   "outputs": [],
   "source": [
    "# broken_axis_box(N2_dna_tes, 'N2', 10000, 90000, 'DNA ', 'DNA', dna_order, True, False)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 6.6.3. <a id='toc6_6_3_'></a>[FX: `make_boxplot`](#toc0_)\n",
    "- make a boxplot of the TE sizes of TEs in a specific class (DNA or Retro) for one genome"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 94,
   "id": "d1cb137b",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2022-08-07T07:03:07.460854Z",
     "start_time": "2022-08-07T07:03:06.801608Z"
    }
   },
   "outputs": [],
   "source": [
    "dna_order = ['Zator', 'hAT', 'CMC', 'Tc1-Mariner', 'Sola', 'Novosib', 'Helitron', 'MITE']\n",
    "retro_order = ['Gypsy', 'Copia', 'LINE', 'SINE', 'ERV']\n",
    "\n",
    "def make_boxplot(dataframe, order: list, genome: str, title: str, color_palette: list, showmeans: bool, to_save: bool = False):\n",
    "    \"\"\" \n",
    "    data: TE dataframe\n",
    "    order: list, what order to plot the families along the x-axis\n",
    "    genome: str, used in the title and filename\n",
    "    title: str, something like 'DNA ' or 'Retro' to indicate what class of TEs are being plotted, used in the title and filename\n",
    "    color_palette: list, list of colors defined in the define constant variables cell for each family\n",
    "    showmeans: bool, whether or not to include a red dot at the mean TE size of each family\n",
    "    to_save: bool, whether or not to save the graph as a pdf (if False, graph will only be output into the notebook)\n",
    "    \"\"\"\n",
    "\n",
    "    sns.set_palette(sns.color_palette(color_palette))\n",
    "\n",
    "    flierprops = dict(marker='o', fillstyle = 'none')\n",
    "    meanpointprops = dict(marker='D', markerfacecolor='red', markeredgecolor = 'black')\n",
    "\n",
    "    fig, axes = plt.subplots(figsize = (12,8))\n",
    "    sns.boxplot(x = 'Family', y = 'Size', data=dataframe, ax = axes, order = order, fliersize = 5, flierprops = flierprops, showmeans = showmeans, meanprops=meanpointprops)\n",
    "\n",
    "    # axes.legend(handles=handles)\n",
    "    axes.title.set_text(f'Size Distributions of {genome} {title}TEs')\n",
    "\n",
    "    if to_save == True:\n",
    "        plt.savefig(f'TEUlt_{genome}_{title}_sizes_box.pdf', dpi=300, facecolor = 'w', edgecolor = 'w', format='pdf')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 97,
   "metadata": {},
   "outputs": [],
   "source": [
    "# make_boxplot(CB_all_filter_df, retro_order, 'CB', 'retro', retro_color_list, True, False)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 6.6.4. <a id='toc6_6_4_'></a>[FX: `make_split_violin`](#toc0_)\n",
    "- makes a split violin plot to compare the distribution of TE sizes by specific family between two genomes\n",
    "- this function uses the seaborn violinplot function, so all of those parameters can be altered within the function"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 114,
   "id": "fb00febe",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2022-08-07T22:05:00.115207Z",
     "start_time": "2022-08-07T22:04:58.964046Z"
    },
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "def make_split_violin(strain_one: str, strain_two: str, te_type: str, to_save: bool = False):\n",
    "    \"\"\" \n",
    "    strain_one: str, name of the first strain to be compared, needs to be the exact string used to label genomes in your dataframes\n",
    "    strain_two: str, name of the second strain to be compared, needs to be the exact string used to label genomes in your dataframes\n",
    "    te_type: str, either 'dna' or 'retro', tells function what TE class to plot\n",
    "    to_save: bool, whether or not to save the graph as a pdf (if False, graph will only be output into the notebook)\n",
    "    \"\"\"\n",
    " \n",
    "    dataframe = all_filter_df[(all_filter_df['Strain'] == strain_one) | (all_filter_df['Strain'] == strain_two)]\n",
    "\n",
    "    if te_type == 'dna':  \n",
    "        order = dna_order\n",
    "        \n",
    "    \n",
    "    elif te_type == 'retro':\n",
    "        order = retro_order\n",
    "        \n",
    "\n",
    "    sns.set_style('white')\n",
    "\n",
    "    fig, ax = plt.subplots(figsize = (13, 8))\n",
    "    \n",
    "    ax = sns.violinplot(x = 'Family', \n",
    "                       y = 'Size', \n",
    "                       data = dataframe, \n",
    "                       order = order, \n",
    "                       inner = 'quartile', \n",
    "                       linewidth = 1, \n",
    "                       scale='width', \n",
    "                       hue = 'Strain', \n",
    "                       split = True,\n",
    "                       palette = colors,\n",
    "                       cut=0,\n",
    "                       alpha=1)\n",
    "    \n",
    "    ax.set_ylabel('Size', fontsize=18)\n",
    "    ax.set_xlabel('Family', fontsize=18)\n",
    "    plt.tick_params(axis='both', which='major', labelsize=16)\n",
    "    plt.legend(bbox_to_anchor=(1.005, 1), loc='upper left',fontsize=14)\n",
    "    plt.tight_layout()\n",
    "\n",
    "    if to_save:\n",
    "        plt.savefig(f'{te_type}_violin_{strain_one}_{strain_two}.pdf', dpi=300, facecolor = 'w', edgecolor = 'w', format='pdf')\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "5d9bd144",
   "metadata": {},
   "outputs": [],
   "source": [
    "# make_split_violin('N2', 'CB', 'dna', False)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 6.6.5. <a id='toc6_6_5_'></a>[FX: `make_box_allgen`](#toc0_)\n",
    "- makes a paired box plot of the size distributions of a TE class for all genomes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 116,
   "id": "812d2695",
   "metadata": {},
   "outputs": [],
   "source": [
    "def make_box_allgen(te_type: str, to_save: bool = False):\n",
    "    \"\"\" \n",
    "    te_type: str, one of either 'dna' or 'retro', tells function what TE class to plot\n",
    "    to_save: bool, whether or not to save the graph as a pdf (if False, graph will only be output into the notebook)\n",
    "    \"\"\"\n",
    "\n",
    "    dataframe = all_filter_df\n",
    "\n",
    "    if te_type == 'dna':  \n",
    "        color_palette = dna_color_list\n",
    "        order = dna_order\n",
    "        title = 'DNA '\n",
    "        filestr = 'dna'\n",
    "    \n",
    "    elif te_type == 'retro':\n",
    "        color_palette = retro_color_list\n",
    "        order = retro_order\n",
    "        title = 'Retro-'\n",
    "        filestr = 'retro'\n",
    "\n",
    "    \n",
    "    flierprops = dict(marker='o', fillstyle = 'none')\n",
    "    meanpointprops = dict(marker='D', markerfacecolor='red', markeredgecolor = 'black')\n",
    "    sns.set_palette(sns.color_palette(color_palette))\n",
    "\n",
    "    fig, axes = plt.subplots(figsize = (12,8))\n",
    "    sns.boxplot(x = 'Family', \n",
    "                y = 'Size', \n",
    "                hue = 'Strain',\n",
    "                palette = colors,\n",
    "                data=dataframe, \n",
    "                ax = axes, \n",
    "                order = order, \n",
    "                fliersize = 5, \n",
    "                flierprops = flierprops, \n",
    "                showmeans = True, \n",
    "                meanprops=meanpointprops)\n",
    "\n",
    "    # axes.legend(handles=handles)\n",
    "    axes.title.set_text(f'Size Distributions of {title}TEs')\n",
    "    plt.legend(bbox_to_anchor=(1.005, 1), loc='upper left',fontsize=14)\n",
    "    plt.tight_layout()\n",
    "    \n",
    "    if to_save:\n",
    "        plt.savefig(f'{filestr}_box_allgenomes.pdf', dpi=300, facecolor = 'w', edgecolor = 'w', format='pdf')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 118,
   "id": "c0065f8e",
   "metadata": {},
   "outputs": [],
   "source": [
    "# make_box_allgen('dna', to_save = False)"
   ]
  }
 ],
 "metadata": {
  "_draft": {
   "nbviewer_url": "https://gist.github.com/358760d4a6cd6836d31ee39d40c62010"
  },
  "gist": {
   "data": {
    "description": "Desktop/labfiles/reasonaTE_N2_out_final/TEUlt_reasonaTE_gff_USE_4_LTRRet_smalldf.ipynb, remaking plots with the size filtered data",
    "public": false
   },
   "id": "358760d4a6cd6836d31ee39d40c62010"
  },
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.5"
  },
  "toc": {
   "base_numbering": 1,
   "nav_menu": {},
   "number_sections": true,
   "sideBar": false,
   "skip_h1_title": false,
   "title_cell": "Table of Contents",
   "title_sidebar": "Contents",
   "toc_cell": true,
   "toc_position": {
    "height": "306.717px",
    "left": "1568.67px",
    "top": "110.2px",
    "width": "255.333px"
   },
   "toc_section_display": true,
   "toc_window_display": true
  },
  "toc-autonumbering": true,
  "toc-showtags": false,
  "varInspector": {
   "cols": {
    "lenName": 16,
    "lenType": 16,
    "lenVar": 40
   },
   "kernels_config": {
    "python": {
     "delete_cmd_postfix": "",
     "delete_cmd_prefix": "del ",
     "library": "var_list.py",
     "varRefreshCmd": "print(var_dic_list())"
    },
    "r": {
     "delete_cmd_postfix": ") ",
     "delete_cmd_prefix": "rm(",
     "library": "var_list.r",
     "varRefreshCmd": "cat(var_dic_list()) "
    }
   },
   "oldHeight": 441.217,
   "position": {
    "height": "40px",
    "left": "1224px",
    "right": "20px",
    "top": "110px",
    "width": "350px"
   },
   "types_to_exclude": [
    "module",
    "function",
    "builtin_function_or_method",
    "instance",
    "_Feature"
   ],
   "varInspector_section_display": "none",
   "window_display": false
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
